---
title: "Diet by Size"
subtitle: "run date: `r format(Sys.time(), '%B %d, %Y')`"
author: "M Fisher"
date: 'written 2022-9-19'
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description

Is diet variability correlated with instar size?

https://rachaellappan.github.io/16S-analysis/correlation-between-otus-with-sparcc.html
https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html

or, Kelly paper with weighted correlation network analysis https://www.nature.com/articles/s41467-019-14105-1#code-availability

In order to directly associate specific taxa to environmental variables (Fig. 3), we used sparse partial least square (sPLS) analysis from the R package mixOmics57,58. We applied the sPLS in regression mode, which will model a causal relationship between the lineages and the environmental traits, that is, PLS will predict environmental traits (e.g., temperature) from lineage abundances. This approach enabled us to identify strong correlations between certain taxa and environmental variables without taking into account the global structure of the planktonic community.



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(vegan)
library(sandwich)
library(MASS)
library(lmtest)
library(labdsv)
library(magrittr)
# library(devtools)
# devtools::install_github("mixOmicsTeam/mixOmics")
library(mixOmics)

source(here('R','eDNA_index.r'))

# User inputs
blastdir   <- 'data/blast'
resultsdir <- 'data/results'
run.nums <- c(1,2)
marker  <- 'lerayXT'

site_names <- data.frame(site=c("KAY","KAY-shell","MARPT","PBNERR","SAMI","SAMT","LAR"),
                         site_name=c("Kayak Pt 1","Kayak Pt 2","March Pt",
                                   "NERR","Samish Isl","Samish Bay","Larrabee"))
```

```{r data, include=FALSE}
## lab metadata (CH,CW,fullness) ##
lab_metadat <- read_csv(here('data','metadata','crab_metadat.csv'))

## filtered taxa ##
dat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_uniqueTaxa.csv")))

dat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## too small taxa (aka "detritus") ##
smdat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_SmallTaxa_uniqueTaxa.csv")))  

smdat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))
```


## Prep data

First, create a data frame with only the carapace width data, per technical replicate. 
```{r}
cw.techs.df <- lab_metadat %>% dplyr::select(sample_id,CW_mm) %>%
  rename(crab_id=sample_id) %>%
  left_join(bind_rows(dat,smdat) %>% dplyr::select(crab_id,sample_id,tech) %>% distinct,
            by=c("crab_id")) %>%
  filter(!is.na(tech))

cw.df <- lab_metadat %>% dplyr::select(sample_id,CW_mm) %>%
  rename(crab_id=sample_id) %>%
  left_join(bind_rows(dat,smdat) %>% dplyr::select(crab_id) %>% distinct() %>% mutate(presence=1),
            by=c("crab_id")) %>%
  filter(!is.na(presence)) %>% dplyr::select(-presence)
```


Then, calculate the eDNA index. I'm going to use all technical replicates.

**large taxa only**
```{r}
crab.lreads.mat <- dat %>%
  group_by(sample_id, new_taxon) %>% summarise(techReads=sum(nReads)) %>%
  pivot_wider(id_cols="sample_id", names_from="new_taxon",values_from="techReads", values_fill=0) %>%
  column_to_rownames("sample_id")

index.lg.mat <- t(eDNA_index(x=t(crab.lreads.mat)))

index.lg.df <- as.data.frame(t(index.lg.mat)) %>%
  rownames_to_column("sample_id") %>% 
  pivot_longer(cols=2:(dim(index.lg.mat)[1]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(sample_id,crab_id,tech,site,estuary) %>% distinct(),by="sample_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```


**micro taxa only**
```{r}
crab.smreads.mat <- smdat %>%
  group_by(sample_id, taxon) %>% summarise(techReads=sum(nReads)) %>%
  pivot_wider(id_cols="sample_id", names_from="taxon",values_from="techReads", values_fill=0) %>%
  column_to_rownames("sample_id")

index.sm.mat <- t(eDNA_index(x=t(crab.smreads.mat)))
```


**all taxa**
```{r}
crab.reads.mat <- bind_rows(smdat,
                            dat %>% dplyr::select(-taxon) %>% rename(taxon=new_taxon)) %>%
  group_by(sample_id, taxon) %>% summarise(techReads=sum(nReads)) %>%
  pivot_wider(id_cols="sample_id", names_from="taxon",values_from="techReads", values_fill=0) %>%
  column_to_rownames("sample_id")

index.mat <- t(eDNA_index(x=t(crab.reads.mat)))

index.df <- as.data.frame(index.mat) %>%
  rownames_to_column("sample_id") %>% 
  pivot_longer(cols=2:(dim(index.mat)[2]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(sample_id,crab_id,tech,site,estuary) %>% distinct(),by="sample_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```

check to make sure all technical replicates are in both data frames
```{r}
all(rownames(index.mat) %in% cw.techs.df$sample_id)
```

For all taxa, also sum up the technical replicates, divide by the number of technical replicates, and calculate the eDNA index once per crab.
```{r}
crabID.reads.mat <- bind_rows(smdat,
                            dat %>% dplyr::select(-taxon) %>% rename(taxon=new_taxon)) %>%
  group_by(crab_id, taxon) %>% summarise(crabReads=sum(nReads),ntechs=length(unique(sample_id))) %>%
  mutate(crabReads_std=crabReads/ntechs) %>% 
  dplyr::select(-crabReads,ntechs) %>%
  pivot_wider(id_cols="crab_id", names_from="taxon",values_from="crabReads_std", values_fill=0) %>%
  column_to_rownames("crab_id")

index.mat2 <- t(eDNA_index(x=t(crabID.reads.mat)))

index.df2 <- as.data.frame(index.mat2) %>%
  rownames_to_column("crab_id") %>% 
  pivot_longer(cols=2:(dim(index.mat2)[2]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(crab_id,site,estuary) %>% distinct(),by="crab_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```




## Something basic

Is there a correlation between the Bray-Curtis distance and the difference in carapace width between every pair of samples? 

Here is what the distribution of carapace widths looks like (red line = gamma dist)
```{r echo=FALSE}
ggplot(cw.df, aes(x=CW_mm)) + geom_histogram(aes(y=after_stat(density))) + theme_bw() +
  stat_function(fun=dgamma, args=list(shape=(mean(cw.df$CW_mm)^2/sd(cw.df$CW_mm)^2), scale=(sd(cw.df$CW_mm)^2/mean(cw.df$CW_mm))), color="red")
```
Calculate the differences in carapace width between each sample. What does the distribution look like?
```{r}
cw_diffs <- abs(outer(cw.df$CW_mm, cw.df$CW_mm, FUN = "-"))
colnames(cw_diffs) <- cw.df$crab_id; rownames(cw_diffs) <- cw.df$crab_id
```

Here are what the pairwise differences look like:
```{r echo=FALSE}
cw_diffs %>%
  as.data.frame() %>%
  rownames_to_column("crab_id") %>%
  pivot_longer(cols=2:(dim(cw_diffs)[2]+1), names_to="crab_id_2",values_to="CW_diff") %>%
  ggplot(aes(x=CW_diff)) + 
  geom_histogram(aes(y=after_stat(density)))
```

Calculate the distances between each sample's prey composition (eDNA index, bray-curtis distance)
```{r}
# get rid of rare prey items that only occur in one crab
index.mat.filter <- dropspc(index.mat2, minocc=2)
index.mat.filter <- index.mat.filter[which(rowSums(index.mat.filter) > 0),]

dim(index.mat2); dim(index.mat.filter)

# calculate bray-curtis distance
index.bc <- vegdist(index.mat.filter, method="bray")
```

Here is what the distance matrix values look like:
```{r echo=FALSE, fig.height=4, fig.width=4}
as.matrix(index.bc) %>%
  as.data.frame() %>%
  rownames_to_column("crab_id") %>%
  pivot_longer(cols=2:(dim(as.matrix(index.bc))[2]+1), names_to="crab_id_2",values_to="BCdistance") %>%
  ggplot(aes(x=BCdistance)) + 
  geom_histogram(aes(y=after_stat(density))) + theme_bw()
```

combine the differences / distances into one data frame.
```{r}
reg.in <- as.matrix(index.bc) %>%
  as.data.frame() %>%
  rownames_to_column("crab_id") %>%
  pivot_longer(cols=2:(dim(as.matrix(index.bc))[2]+1), names_to="crab_id2",values_to="BCdistance") %>%
  left_join(
    cw_diffs %>%
      as.data.frame() %>%
      rownames_to_column("crab_id") %>%
      pivot_longer(cols=2:(dim(cw_diffs)[2]+1), names_to="crab_id2",values_to="CW_diff"),
    by=c("crab_id","crab_id2"))
```

What do they look like plotted against each other?
```{r}
ggplot(reg.in,aes(x=CW_diff,y=BCdistance)) +
  geom_point() + labs(x="Difference in CW (mm)", y="Difference in Prey Composition") +
  theme_bw()
```

Use a generalized linear model to correlate the carapace width differences with the distance values (quasi-binomial, logit-link -- bray-curtis distances are between 0 and 1, left skewed)
```{r}
cw.glm <- glm(BCdistance ~ CW_diff, data = reg.in, family = quasibinomial('logit'))
summary(cw.glm)
```

Address potential heteroskedasticity of errors:
```{r}
robust.se.glm1 <- sqrt(diag(vcovHC(cw.glm , type="HC0")))
coeftest(cw.glm, vcovHC(cw.glm , type="HC0"))
```




## Sparse partial least squares analysis


Kelly et al. :
> In order to directly associate specific taxa to environmental variables (Fig. 3), we used sparse partial least square (sPLS) analysis from the R package mixOmics. We applied the sPLS in regression mode, which will model a causal relationship between the lineages and the environmental traits, that is, PLS will predict environmental traits (e.g., temperature) from lineage abundances. This approach enabled us to identify strong correlations between certain taxa and environmental variables without taking into account the global structure of the planktonic community.




### large taxa

#### all replicates

The carapace width data will be "Y" in the sPLS, and the diet data will be "X"

```{r fig.width=3, fig.height=3}
cw.lg.mat <- as.matrix(cw.techs.df %>% filter(sample_id %in% rownames(index.lg.mat)) %>%
  arrange(match(sample_id, rownames(index.lg.mat))) %>%
  column_to_rownames(var="sample_id") %>% dplyr::select(CW_mm))

hist(cw.lg.mat[,1], main="histogram of carapace width")
```

make sure the rows in the carapace width data exactly match the rows in the diet data. 
```{r}
all(rownames(cw.lg.mat) == rownames(index.lg.mat))
```


Defining the ‘best’ number of dimensions to explain the data requires we first launch a PLS1 model with a large number of components. Some of the outputs from the PLS1 object are then retrieved in the perf() function to calculate the Q2

criterion using repeated 10-fold cross-validation.
```{r eval=FALSE}
cw.lgtax.pls4 <- pls(Y = cw.lg.mat, X = index.lg.mat,  ncomp=4, mode = 'regression')

# test.pls4 <- pls(Y = liver.toxicity$clinic[,"ALB.g.dL."], X = liver.toxicity$gene,  ncomp=4, mode = 'regression')

set.seed(33)  # For reproducibility with this handbook, remove otherwise
Q2.lgtax.pls4 <- perf(cw.lgtax.pls4, validation = 'Mfold', 
                      folds = 10, nrepeat = 5)
plot(Q2.lgtax.pls4, criterion = 'Q2')
```
Error in X.test %*% a.cv : non-conformable arguments (from `perf`)


```{r eval=FALSE}
out.list <- list(index.lg.mat, cw.lg.mat); names(out.list) <- c("X","Y")
saveRDS(out.list, here('data','results','test_spls.rds'))
rm(out.list)
readRDS(here('data','results','test_spls.rds'))


hist(index.lg.mat, main="histogram of eDNA index (X)")
```


#### crab only

Try again, for crab instead of replicates. The carapace width data will be "Y" in the sPLS, and the diet data will be "X"

make the "X" data frame by crab, instead of by replicate. Do this by summing reads across replicates, dividing by the number of replicates in which that taxon was found, and then calculating the eDNA index. 
```{r}
crabID.lreads.mat <- dat %>%
  group_by(crab_id, taxon) %>% summarise(crabReads=sum(nReads),ntechs=length(unique(sample_id))) %>%
  mutate(crabReads_std=crabReads/ntechs) %>% 
  dplyr::select(-crabReads,ntechs) %>%
  pivot_wider(id_cols="crab_id", names_from="taxon",values_from="crabReads_std", values_fill=0) %>%
  column_to_rownames("crab_id")

index.lg.mat2 <- t(eDNA_index(x=t(crabID.lreads.mat)))

index.lg.df2 <- as.data.frame(t(index.lg.mat2)) %>%
  rownames_to_column("crab_id") %>% 
  pivot_longer(cols=2:(dim(index.lg.mat2)[1]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(crab_id,site,estuary) %>% distinct(),by="crab_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```


```{r fig.width=3, fig.height=3}
cw.lg.mat <- as.matrix(cw.df %>% filter(crab_id %in% rownames(index.lg.mat2)) %>%
  arrange(match(crab_id, rownames(index.lg.mat2))) %>%
  column_to_rownames(var="crab_id") %>% dplyr::select(CW_mm))

hist(cw.lg.mat[,1], main="histogram of carapace width")
```

make sure the rows in the carapace width data exactly match the rows in the diet data. 
```{r}
all(rownames(cw.lg.mat) == rownames(index.lg.mat2))
```


Defining the ‘best’ number of dimensions to explain the data requires we first launch a PLS1 model with a large number of components. Some of the outputs from the PLS1 object are then retrieved in the perf() function to calculate the Q2

criterion using repeated 10-fold cross-validation.
```{r eval=FALSE}
cw.lgtax.pls4 <- pls(Y = cw.lg.mat, X = index.lg.mat2,  ncomp=4, mode = 'regression')

# test.pls4 <- pls(Y = liver.toxicity$clinic[,"ALB.g.dL."], X = liver.toxicity$gene,  ncomp=4, mode = 'regression')

set.seed(33)  # For reproducibility with this handbook, remove otherwise
Q2.lgtax.pls4 <- perf(cw.lgtax.pls4, validation = 'Mfold', 
                      folds = 10, nrepeat = 5)
plot(Q2.lgtax.pls4, criterion = 'Q2')
```
Error in X.test %*% a.cv : non-conformable arguments (from `perf`)



## Clustering




