---
title: "Dungeness Diet Results 1"
subtitle: "run date: `r format(Sys.time(), '%B %d, %Y')`"
author: "M Fisher"
date: 'written 2022-12-03'
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description


Summary document to produce major results information for text, tables, figures. Includes:

1. Metadata: 

  - How many crab were sequenced, total and per site?
  - How many sequenced crab had putative prey in their stomach contents?
  - What were the mean carapace widths per site?

2. Prey diversity metrics:

  - alpha diversity - species and taxa, overall and per site
  - alpha diversity x stomach fullness 
  - beta diversity - What was the unique number of taxa identified per per site?
  - taxonomic distinctness (vegan doc)[https://search.r-project.org/CRAN/refmans/vegan/html/taxondive.html]
  - species richness within each phyla

3. Analyze site-by-site differences  

  - Site differences in presence / absence: PERMANOVA & PERMDISP 
  - Site differences in eDNA index: PERMANOVA & PERMDISP  
  - PERMDISP as a PCoA **Main text figure**

4. Capture Prey identifications

 - Frequency of Occurrence for prey groups **Main text figure**


Each of the following is completed for unique taxa (any level allowed in the final filtered dataset), and then again for a species-only data set.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(here)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(vegan)
library(zetadiv)
library(magrittr)
library(ggordiplots)
library(grid); library(gridExtra)
source(here('R','squash_axis.r'))
source(here('R','ggordiplot_custom.r'))
source(here('R','eDNA_index.r'))
# User inputs
blastdir   <- 'data/blast'
resultsdir <- 'data/results'
run.nums <- c(1,2)
marker  <- 'lerayXT'

estuary_colors <- data.frame(estuary=c("Port Susan Bay","Padilla Bay", "Samish Bay"),
                             key=c("#33a02c","#1f78b4","chocolate3"))

site_names <- data.frame(site=c("KAY","KAY-shell","MARPT","PBNERR","SAMI","SAMT","LAR"),
                         site_name=c("Kayak Pt 1","Kayak Pt 2","March Pt",
                                   "NERR","Samish Isl","Samish Bay","Larrabee"))
```
```{r data, include=FALSE}
## lab metadata (CH,CW,fullness) ##
lab_metadat <- read_csv(here('data','metadata','crab_metadat.csv'))

## seq sample id metadata ##
run_metadat_files <- c('data/DCRB_Run1_samples.csv','data/DCRB_Run2_samples.csv')
for(i in seq(1,length(run_metadat_files))){
  tmp.meta <- read_csv(here(run_metadat_files[i])) %>% mutate(MiSeqRun=run.nums[i]) %>% mutate(crab_num=as.character(crab_num), tech=as.character(tech))
  if(i==1){    metadat <- tmp.meta  } else{
    metadat <- bind_rows(metadat,tmp.meta)
  } }

## blast output, pre-filter ##
blastout <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_blast_lca_taxonomy.csv")))

blastout %<>% dplyr::select(-mi_seq_run,-sample_id) %>%
  left_join(metadat, by=c("Sample_name","Locus"="locus","MiSeqRun")) %>%
  rename(sample_id=sample_label) %>%
  mutate(crab_num=as.numeric(crab_num)) %>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## filtered taxa ##
dat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_uniqueTaxa.csv")))

dat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## too small taxa (aka "detritus") ##
smdat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_SmallTaxa_uniqueTaxa.csv")))  

smdat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## combine all taxa ##
alldat <- bind_rows(dat,smdat)
```
<br>

# 1. Metadata

### Sample Sizes

Total number of crab with prey info: 
```{r}
length(alldat %>% pull(crab_id) %>% unique())
```

What are the final sample sizes for this data set?
```{r}
alldat %>%
  group_by(estuary,site) %>% summarise(`Crab count (all)`=length(unique(crab_id))) %>%
  left_join(smdat %>% 
              group_by(estuary,site) %>% summarise(`Crab count (small)`=length(unique(crab_id)))) %>%
  left_join(dat %>%
  group_by(estuary,site) %>% summarise(`Crab count`=length(unique(crab_id))))
```
<br>

What % of crab sequenced had putative prey (without single-celled organisms) in their stomach contents?
```{r}
pcrab_with_preytax <- blastout %>% dplyr::select(site,crab_num,estuary) %>% distinct() %>%
  mutate(crab_num=as.character(crab_num)) %>%
  left_join(dat %>% dplyr::select(crab_num,site,estuary) %>% 
              distinct() %>% mutate(with_prey="1", crab_num=as.character(crab_num))) %>%
  mutate(with_prey=ifelse(is.na(with_prey),"0",with_prey)) %>%
  group_by(estuary,site) %>%
  mutate(total_crab=length(unique(crab_num))) %>%
  group_by(estuary,site,total_crab,with_prey) %>%
  summarise(pcrab=length(unique(crab_num))/total_crab) %>% distinct() %>%
  pivot_wider(names_from=with_prey,values_from=pcrab)
pcrab_with_preytax 
```
What % of crab sequenced had putative prey **species** (without single-celled organisms) in their stomach contents?
```{r}
blastout %>% dplyr::select(site,crab_num,estuary) %>% distinct() %>%
  mutate(crab_num=as.character(crab_num)) %>%
  left_join(dat %>% filter(rank=="species") %>% dplyr::select(crab_num,site,estuary) %>% 
              distinct() %>% mutate(with_prey="1", crab_num=as.character(crab_num))) %>%
  mutate(with_prey=ifelse(is.na(with_prey),"0",with_prey)) %>%
  group_by(estuary,site) %>%
  mutate(total_crab=length(unique(crab_num))) %>%
  group_by(estuary,site,total_crab,with_prey) %>%
  summarise(pcrab=round(length(unique(crab_num))/total_crab, 3)) %>% distinct() %>%
  pivot_wider(names_from=with_prey,values_from=pcrab)
```


```{r}
write_csv(pcrab_with_preytax, here(resultsdir, 'lerayXT_runs1-2_percent_crab_with_prey_taxa.csv'))
```


*Important for further analysis* - do we have the same number of technical replicates per crab?
```{r}
with(alldat %>% group_by(crab_id) %>% summarise(ntech=length(unique(tech))), table(ntech))
with(dat %>% group_by(crab_id) %>% summarise(ntech=length(unique(tech))), table(ntech))
with(smdat %>% group_by(crab_id) %>% summarise(ntech=length(unique(tech))), table(ntech))
```

ntech
 2  3 
 6 52 
ntech
 1  2  3 
 5  5 46 
ntech
 1  2  3 
 4  8 42 
 
crab with only two replicates in the full data set:
KAY_05
KAY_23
MARPT_12
SAMT_02
SIN_02
SIN_08

So, there are enough crab with < 3 technical replicates that when running analyses like the PERMANOVA, I want to either (1) sum reads across technical replicates, or (2) nest technical replicates within crabs, like factors.

### Carapace width

```{r}
crab_metadat <- alldat %>% dplyr::select(site,estuary,crab_id) %>% distinct() %>%
  left_join(lab_metadat,by=c("crab_id"="sample_id","estuary"))

all(!is.na(crab_metadat$CW_mm))
```

```{r}
crab_metadat %>% group_by(site) %>% summarise(mean_cw=round(mean(CW_mm), 2), sdev_cw=round(sd(CW_mm), 2),
                                              mean_ch=round(mean(CH_mm), 2), sdev_ch=round(sd(CH_mm), 2))
```

Graph CW x CH, and save as an object for figure 1.
```{r fig.height=3}
fig1b <- ggplot(crab_metadat, aes(x=CW_mm,y=CH_mm,color=fct_relevel(estuary, estuary_colors$estuary))) + 
  geom_point() + facet_grid(cols=vars(estuary)) +
  scale_color_manual(values=estuary_colors$key, name="estuary") +
  labs(x="Carapace width (mm)", y="Carapace height (mm)") +
  theme_bw() + theme(legend.position="none", strip.background=element_rect("white"))
fig1b

if(!knitting){saveRDS(fig1b, here('figs','fig1b_obj.rds'))}
```

Graph mean and sdev carapace width and height for each site, and save for supplement.
```{r}
png(here('figs','figS_CW_CH_bysite.png'),res=300,height=1700,width=1900)
  
crab_metadat %>% left_join(site_names) %>%
  group_by(estuary,site,site_name) %>%
  summarise(mean_cw=mean(CW_mm), sdev_cw=sd(CW_mm),
                                              mean_ch=mean(CH_mm), sdev_ch=sd(CH_mm)) %>%
  mutate(site_name=factor(site_name, levels=c("Kayak Pt 1","Kayak Pt 2","March Pt","NERR","Samish Isl","Samish Bay","Larrabee"))) %>%
ggplot(aes(x=mean_cw,y=mean_ch,pch=site_name,color=site_name)) + 
  geom_point() + 
  geom_errorbar(aes(ymin=mean_ch-sdev_ch,ymax=mean_ch+sdev_ch),linewidth=0.25) +
  geom_errorbarh(aes(xmin=mean_cw-sdev_ch,xmax=mean_cw+sdev_cw),linewidth=0.25) +
  scale_color_manual(values=c("darkgreen","chartreuse3","#1f78b4","cadetblue2","chocolate4","chocolate2","orange2"),name="site") +
  scale_shape_manual(values=c(15,15,16,16,17,17,17),name="site") +
  labs(x="Carapace width (mm)", y="Carapace height (mm)") +
  xlim(c(5.5,15)) + ylim(c(5.5,12)) +
  theme_bw() 

dev.off()
```



# Taxa

Allows for non-species-level IDs, as long as they are unique. 

What % of the unique taxa we've identified go down to the species level, for **all data**?
```{r}
# overall, unique taxa observed
alldat_overall <- get_unique_taxa(alldat)
# percent at each rank?
with(alldat_overall %>% dplyr::select(taxon,rank) %>% distinct(),table(rank)) / length(alldat_overall %>% pull(taxon) %>% unique())
```

What % of the unique taxa we've identified go down to the species level, for data **excluding single-celled organisms**?
```{r}
# overall, unique taxa observed
dat_overall <- get_unique_taxa(dat)
# percent at each rank?
with(dat_overall %>% dplyr::select(taxon,rank) %>% distinct(),table(rank)) / length(dat_overall %>% pull(taxon) %>% unique())
```
What % of the unique taxa we've identified go down to the species level, for **only single-celled organisms**?
```{r}
# overall, unique taxa observed
smdat_overall <- get_unique_taxa(smdat)
# percent at each rank?
with(smdat_overall %>% dplyr::select(taxon,rank) %>% distinct(),table(rank)) / length(smdat_overall %>% pull(taxon) %>% unique())
```

## eDNA Index

calculate eDNA index for all technical replicates (1 row = 1 crab x 1 replicate)
```{r}
sample.reads.mat <- dat %>%
  group_by(sample_id, taxon) %>% summarise(techReads=sum(nReads)) %>%
  pivot_wider(id_cols="sample_id", names_from="taxon",values_from="techReads", values_fill=0) %>%
  column_to_rownames("sample_id")

index2.mat <- eDNA_index(x=t(sample.reads.mat))

index2.df <- as.data.frame(t(index2.mat)) %>%
  rownames_to_column("sample_id") %>% 
  pivot_longer(cols=2:(dim(index2.mat)[1]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(sample_id,crab_id,tech,site,estuary) %>% distinct(),by="sample_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```

calculate eDNA index from the sum of reads across technical replicates (1 row = 1 crab)
```{r}
crab.reads.mat <- dat %>%
  group_by(crab_id, taxon) %>% summarise(sumReads=sum(nReads)) %>%
  pivot_wider(id_cols="crab_id", names_from="taxon",values_from="sumReads", values_fill=0) %>%
  column_to_rownames("crab_id")

index.mat <- eDNA_index(x=t(crab.reads.mat))

index.df <- as.data.frame(t(index.mat)) %>%
  rownames_to_column("crab_id") %>% 
  pivot_longer(cols=2:(dim(index.mat)[1]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(crab_id,site,estuary) %>% distinct(),by="crab_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```



## 2. Prey Diversity

### Alpha div

What is the total number of unique taxa observed, without unicellular species? **99**  
```{r}
dat_overall <- get_unique_taxa(dat)
length(dat_overall %>% pull(taxon) %>% unique())
```

What is the total number of unique taxa observed among unicellular algae / diatoms / rotifers / protists? **38**
```{r}
smdat_overall <- get_unique_taxa(smdat)
length(smdat_overall %>% pull(taxon) %>% unique())
```
How many phyla does this represent?
```{r}
smdat_overall %>% mutate(level=ifelse(!is.na(phylum), "phylum", "class"),
                         to_count=ifelse(!is.na(phylum), phylum, class)) %>% group_by(level) %>% summarise(n=length(unique(to_count)))
```

What is the overall number of unique taxa? **136**
```{r}
length(alldat_overall %>% pull(taxon) %>% unique())
```
How many phyla does this represent?
```{r}
dat_overall %>% mutate(level=ifelse(!is.na(phylum), "phylum", "class"),
                         to_count=ifelse(!is.na(phylum), phylum, class)) %>% group_by(level) %>% summarise(n=length(unique(to_count)))
```

#### per site

What is the taxonomic richness ($\alpha$ diversity) at each site, **for all taxa**? Standardized site richness is simply the  richness / total crab from site.
```{r}
dat_sites <- get_unique_taxa(taxa.df=alldat, level="site")

richness.site <- dat_sites %>% group_by(site) %>% summarise(richness=length(unique(taxon)),
                                                            ncrab=length(unique(crab_id))) %>%
  mutate(std.richness=richness/ncrab)
richness.site
```
<br>

What is the taxonomic richness ($\alpha$ diversity) at each site, **excluding single-celled organisms**? Standardized site richness is simply the  richness / total crab from site.
```{r}
dat_sites <- get_unique_taxa(taxa.df=dat, level="site")

richness.site <- dat_sites %>% group_by(site) %>% summarise(richness=length(unique(taxon)),
                                                            ncrab=length(unique(crab_id))) %>%
  mutate(std.richness=richness/ncrab)
richness.site
```


#### per crab


what is the taxonomic richness ($\alpha$ diversity) per crab?
```{r message=FALSE}

allrichness.crab <- alldat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>%
  left_join(estuary_colors,by="estuary")

paste0("overall taxonomic richness per crab ranges from ", min(allrichness.crab$richness), " to ", max(allrichness.crab$richness), " with a median of ", median(allrichness.crab$richness))


richness.crab <- dat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>%
  left_join(estuary_colors,by="estuary")

paste0("taxonomic richness per crab ranges from ", min(richness.crab$richness), " to ", max(richness.crab$richness), " with a median of ", median(richness.crab$richness))


smrichness.crab <- smdat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>%
  left_join(estuary_colors,by="estuary")

paste0("unicellular taxonomic richness per crab ranges from ", min(smrichness.crab$richness), " to ", max(smrichness.crab$richness), " with a median of ", median(smrichness.crab$richness))
```
```{r}
plot.alpha <- allrichness.crab %>% mutate(dataset="All Taxa") %>%
  bind_rows(richness.crab %>% mutate(dataset="Excluding Single-celled organisms")) %>%
  bind_rows(smrichness.crab %>% mutate(dataset="Single-celled organisms alone")) %>%
  left_join(site_names, by="site") %>%
  mutate(site_name=factor(site_name, levels=c("Kayak Pt 1","Kayak Pt 2","March Pt","NERR","Samish Isl","Samish Bay", "Larrabee"))) %>%
  ggplot(aes(x=site_name,y=richness,color=estuary)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  facet_grid(cols=vars(dataset)) +
  labs(y="Number Taxa per crab",x="Sampling Site") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  scale_color_manual(values=unique(richness.crab$key)) +
  theme_bw() + theme(axis.text.x=element_text(angle=25, hjust=1))

plot.alpha
```
<br>

Save plot.
```{r echo=TRUE}
png(here('figs','figS_alpha-div.png'), res=300, height=1000,width=2800)
plot.alpha
dev.off()
```
<br>


How is ($\alpha$ diversity) distributed across taxonomic levels?
```{r message=FALSE}
# richness.phyla.site <- dat_sites %>%
#   mutate(group=ifelse(!is.na(phylum),phylum,class)) %>%
#   group_by(estuary,site,group) %>%
#   summarise(richness=length(unique(taxon)))  %>%
#   left_join(site_names)

richness.phyla <- alldat_overall %>%
  mutate(group=ifelse(!is.na(phylum),phylum,class)) %>%
  group_by(group) %>%
  summarise(richness=length(unique(taxon))) %>%
  arrange(richness)

# richness.phyla <- richness.phyla %>%
#   mutate(estuary="Overall",site="Overall") %>%
#   bind_rows(richness.phyla.site)
```

```{r fig.height=5,fig.width=5}
richness.phyla %>%
  ggplot(aes(y=fct_reorder(group, richness),x=richness)) +
  geom_col(position="dodge") +
  labs(x="Richness",y="Taxonomic Group") +
  # scale_fill_manual(values=c("cyan4","darkblue","darkorange","coral3","gray30")) +
  theme_bw()
```

<br>
save as a table
```{r}
if(!knitting){
write_csv(richness.phyla, here('data','results','lerayXT_r1-2_tax-richness-overall.csv'))
}
```


#### Alpha div x Stomach Fullness

Consider alpha diversity with / without the 'detritus' taxa, and then 'detritus' taxa alone
```{r message=FALSE}
### without detritus ###
richness.crab <- richness.crab %>%
  left_join(crab_metadat, by=c("crab_id","estuary"))

### with detritus ###
drichness.crab <- smdat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>% 
  left_join(estuary_colors,by="estuary") %>%
  left_join(crabdat)

### combined ###
all.richness.crab <- alldat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>% 
  left_join(estuary_colors,by="estuary") %>%
  left_join(crabdat)
```


```{r}
p.alpha3 <- ggplot(richness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="\nOnly Multi-cellular Taxa") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha2 <- ggplot(all.richness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="\nAll Taxa") +
  scale_y_continuous(limits=c(0,max(all.richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha4 <- ggplot(drichness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="Only Single-celled\n Organisms") +
  scale_y_continuous(limits=c(0,max(drichness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()

y.grob <- textGrob("Number Taxa per Crab", 
                   gp=gpar(col="black"), rot=90)

x.grob <- textGrob("Stomach Fullness", 
                   gp=gpar(col="black"))


main <- plot_grid(p.alpha2,p.alpha3,p.alpha4, ncol=3)
alpha_by_fll <- plot_grid(left=y.grob,main,NULL,bottom=x.grob, rel_heights=c(1,0.05), rel_widths=c(0.05,1))
alpha_by_fll
```
<br>

Save plot with the alpha diversity x site plot
```{r echo=TRUE}
# png(here('figs','figS_alpha-div_site_fullness.png'), res=200, height=1000,width=1500)
# plot_grid(plot.alpha + theme(plot.margin=margin(t=0,r=0,b=10,l=20)), alpha_by_fll, nrow=2)
# dev.off()

png(here('figs','figS_alpha-div_fullness.png'), res=200, height=800,width=1500)
alpha_by_fll
dev.off()
```
<br>
<br>

#### Alpha Div x Size

```{r}
p.alpha2 <- ggplot(richness.crab %>% filter(!is.na(CW_mm)), aes(x=CW_mm,y=richness)) +
  geom_point() +
  labs(y="",x="", title="\nOnly Multi-cellular Taxa") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha1 <- ggplot(all.richness.crab %>% filter(!is.na(CW_mm)), aes(x=CW_mm,y=richness)) +
  geom_point() +
  labs(y="",x="", title="\nAll Taxa") +
  scale_y_continuous(limits=c(0,max(all.richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha3 <- ggplot(drichness.crab %>% filter(!is.na(CW_mm)), aes(x=CW_mm,y=richness)) +
  geom_point() +
  labs(y="",x="", title="Only Single-celled\n Organisms") +
  scale_y_continuous(limits=c(0,max(drichness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()

y.grob <- textGrob("Number Taxa per Crab", 
                   gp=gpar(col="black"), rot=90)

x.grob <- textGrob("Carapace Width (mm)", 
                   gp=gpar(col="black"))


main <- plot_grid(p.alpha1,p.alpha2,p.alpha3, ncol=3)
alpha_by_cw <- plot_grid(left=y.grob,main,NULL,bottom=x.grob, rel_heights=c(1,0.05), rel_widths=c(0.05,1))
alpha_by_cw
```
Save plot
```{r echo=TRUE}
png(here('figs','figS_alpha-div_cw.png'), res=200, height=800,width=1500)
alpha_by_cw
dev.off()
```


### Saturation

#### all taxa

At each site, how does adding one additional crab change alpha diversity?
```{r fig.height=3, fig.width=4}
all_sp <- c()
added_sp <- data.frame(ncrab=as.numeric(),
                       nsp=as.numeric())
for(i in seq(1,length(unique(alldat$crab_id)))){
  new_sp <- alldat %>%
    filter(crab_id==unique(alldat$crab_id)[i]) %>%
    filter(rank=="species") %>%
    dplyr::select(crab_id,species) %>%
    distinct() %>%
    filter(!(species %in% all_sp)) %>%
    pull(species)
    all_sp <- c(all_sp, new_sp)
  added_sp <- added_sp %>% bind_rows(data.frame(ncrab=i, nsp=length(all_sp)))
}

sat.overall.plot <- ggplot(added_sp,aes(x=ncrab,y=nsp)) +
  geom_point(size=0.5) + geom_path() +
  labs(y="total prey taxa",x="number of crab") + theme_classic()
``` 
```{r fig.height=3, fig.width=4}

added_sp_est <- data.frame(estuary=as.character(),
                       ncrab=as.numeric(),
                       nsp=as.numeric())
for(s in unique(alldat$estuary)){
  dat.filter <- filter(alldat, estuary==s)
  all_sp <- c()
  for(i in seq(1,length(unique(dat.filter$crab_id)))){
    new_sp <- dat.filter %>%
      filter(crab_id==unique(dat.filter$crab_id)[i]) %>%
      filter(rank=="species") %>%
      dplyr::select(crab_id,species) %>%
      distinct() %>%
      filter(!(species %in% all_sp)) %>%
      pull(species)
    all_sp <- c(all_sp, new_sp)
    added_sp_est <- added_sp_est %>% bind_rows(data.frame(estuary=s, ncrab=i, nsp=length(all_sp)))
  }
}
```
```{r}
plotdat <- added_sp_est %>% 
  bind_rows(added_sp %>% mutate(estuary="Overall")) %>%
  left_join(estuary_colors %>% bind_rows(data.frame(estuary="Overall",key="gray30")))

ggplot(plotdat, aes(x=ncrab,y=nsp,color=estuary)) +
  geom_point( size=0.5) + geom_path() +
  scale_color_manual(values=c("gray30",estuary_colors$key[c(2,1,3)])) +
  labs(y="total prey taxa",x="number of crab") + theme_classic()
``` 
<br>

#### excluding single-celled organisms

At each site, how does adding one additional crab change alpha diversity?
```{r fig.height=3, fig.width=4}
all_sp <- c()
added_sp <- data.frame(ncrab=as.numeric(),
                       nsp=as.numeric())
for(i in seq(1,length(unique(dat$crab_id)))){
  new_sp <- dat %>%
    filter(crab_id==unique(dat$crab_id)[i]) %>%
    filter(rank=="species") %>%
    dplyr::select(crab_id,species) %>%
    distinct() %>%
    filter(!(species %in% all_sp)) %>%
    pull(species)
    all_sp <- c(all_sp, new_sp)
  added_sp <- added_sp %>% bind_rows(data.frame(ncrab=i, nsp=length(all_sp)))
}

sat.overall.plot <- ggplot(added_sp,aes(x=ncrab,y=nsp)) +
  geom_point(size=0.5) + geom_path() +
  labs(y="total prey taxa",x="number of crab") + theme_classic()
``` 
```{r fig.height=3, fig.width=4}

added_sp_est <- data.frame(estuary=as.character(),
                       ncrab=as.numeric(),
                       nsp=as.numeric())
for(s in unique(dat$estuary)){
  dat.filter <- filter(dat, estuary==s)
  all_sp <- c()
  for(i in seq(1,length(unique(dat.filter$crab_id)))){
    new_sp <- dat.filter %>%
      filter(crab_id==unique(dat.filter$crab_id)[i]) %>%
      filter(rank=="species") %>%
      dplyr::select(crab_id,species) %>%
      distinct() %>%
      filter(!(species %in% all_sp)) %>%
      pull(species)
    all_sp <- c(all_sp, new_sp)
    added_sp_est <- added_sp_est %>% bind_rows(data.frame(estuary=s, ncrab=i, nsp=length(all_sp)))
  }
}
```
```{r}
plotdat2 <- added_sp_est %>% 
  bind_rows(added_sp %>% mutate(estuary="Overall")) %>%
  left_join(estuary_colors %>% bind_rows(data.frame(estuary="Overall",key="gray30")))

ggplot(plotdat2, aes(x=ncrab,y=nsp,color=estuary)) +
  geom_point( size=0.5) + geom_path() +
  scale_color_manual(values=c("gray30",estuary_colors$key[c(2,1,3)])) +
  labs(y="total prey taxa",x="number of crab") + theme_classic()
``` 
<br>

```{r}

p1 <- ggplot(plotdat, aes(x=ncrab,y=nsp,color=estuary)) +
  geom_point( size=0.5) + geom_path() +
  scale_color_manual(values=c("gray30",estuary_colors$key[c(2,1,3)])) +
  labs(y="total prey taxa",x="number of crab",subtitle="all prey taxa") + theme_bw() +
  theme(legend.position="none")
p2 <- ggplot(plotdat2, aes(x=ncrab,y=nsp,color=estuary)) +
  geom_point( size=0.5) + geom_path() +
  scale_color_manual(values=c("gray30",estuary_colors$key[c(2,1,3)])) +
  labs(y="",x="number of crab", subtitle="excluding single-celled organisms") + theme_bw()


png(here('figs','figS_saturation_curve_alltax_lgtax.png'), res=300,height=1300,width=2200)
plot_grid(p1,p2, rel_widths=c(0.7,1))
dev.off()
```


### Distinctness


*Taxonomic distinctness* is the averaged taxonomic distances among species or individuals in the community (Clarke & Warwick 1998, 2001). These figures display the distinctness in prey assemblages between crab at the same site (with presence/absence data, both distinctness and diversity reduce to the same index, $\Delta^{+}$, Clarke & Warwick (1998)). 

Note that this metric can only be run for those species which have their full taxonomy available:
```{r}
tmpn <- length(unique(dat %>% filter(!is.na(phylum) & !is.na(class) & !is.na(order) & !is.na(family) & !is.na(genus)) %>% pull(taxon)))
message('taxonomic distinctness calculated for ', tmpn, ' species; this represents ', round(tmpn/length(unique(dat$taxon)),2)*100, '% of all species in the dataset.')
```
<br>

$\Delta^(+)$ per site:
```{r message=FALSE, warning=FALSE}
# taxonomic distances
sp.taxonomy <- dat %>%
  dplyr::select(phylum,class,order,family,genus,species) %>% distinct() %>%
  rename("Phylum"=phylum, "Class"=class,"Order"=order,"Family"=family,"Genus"=genus) %>%
  filter((!is.na(Phylum) & !is.na(Class) & !is.na(Order)) & !is.na(Family) & !is.na(Genus))
x <- as.data.frame(sp.taxonomy %>% dplyr::select(Genus,Family,Order,Class,Phylum))
rownames(x) <- sp.taxonomy$species
sp.dist.mat <- taxa2dist(x, varstep = FALSE, check = TRUE, labels=rownames(x))

# presence / absence matrix
prey.sp.site.df <- dat %>%
  filter(species %in% sp.taxonomy$species) %>%
  dplyr::select(Site, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="Site", names_from="taxon",values_from="presence", values_fill=0)

prey.sp.site.mat <- as.data.frame(prey.sp.site.df %>% dplyr::select(-Site))
rownames(prey.sp.site.mat) <- prey.sp.site.df$Site


# calculate distance
site.taxdist <- taxondive(comm=prey.sp.site.mat, dis=sp.dist.mat, match.force=TRUE)
site.taxdist.df <- data.frame(site=names(site.taxdist$Species),
                              nsp=site.taxdist$Species,
                              dplus=site.taxdist$Dplus,
                              dplus.sd=site.taxdist$sd.Dplus) %>%
  mutate(site=ifelse(site=="KAY","Kayak\nPoint 1",ifelse(site=="KAY-shell", "Kayak\nPoint 2",
                                                        ifelse(site=="LAR","Larrabee",
                                                               ifelse(site=="MARPT","March\nPoint",NA)))))
```
```{r fig.width=5, fig.height=4}
pd1 <- ggplot(site.taxdist.df, aes(x=site,y=dplus, col=nsp)) +
  geom_point() +
  geom_errorbar(aes(ymin=dplus-dplus.sd,ymax=dplus+dplus.sd), width=0.5) +
  scale_y_continuous(limits=c(75,100)) +
  labs(x="Site",y="Distinctness", title="Taxonomic Distinctness",subtitle="*calculated from 86% of species in data set") +
  scale_color_gradient(low = "plum2", high = "coral4", na.value = NA, name="Species\n Count") +
  theme_bw()

pd1
```

<br>

Save plot.
```{r echo=TRUE}
png(here('figs','fig2_distinctness.png'), res=200, height=600,width=800)
pd1
dev.off()
```
<br>
<br>


<br>
<br>




## Beta & Zeta diversity

Where appropriate, calculate these indices by summing reads / summarizing presence across technical replicates. 

### Beta diversity

What is the $\beta$ diversity for prey taxon? (the ratio between regional and local prey diversity) The greater the similarity in community composition between multiple communities, the lower the value of β-diversity for that set of communities.

I do have very different sample sizes between sites. According to Barwell et al. 2014, "In sampling simulations S1 (Unbiased by undersampling) and S2 (Unbiased by unequal sampling effort ) (Table 1; Figs S12 and S13), most presence–absence metrics were positively biased by undersampling, with the exception of βChao Sørensen and βChao Jaccard which have a correction for undersampling." But those need species abundance data...

First, among all sites *all taxa*
```{r}
prey.sp.site.mat <- alldat %>%
  dplyr::select(site, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="site", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv <- as.matrix(betadiver(prey.sp.site.mat, method="j")) 
colnames(bdiv) <-c(prey.sp.site.mat$site); rownames(bdiv) <- c(prey.sp.site.mat$site)
bdiv
```


among all sites *excluding single-celled organisms*
```{r}
prey.sp.site.mat2 <- dat %>%
  dplyr::select(site, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="site", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv2 <- as.matrix(betadiver(prey.sp.site.mat2, method="j")) 
colnames(bdiv2) <-c(prey.sp.site.mat2$site); rownames(bdiv2) <- c(prey.sp.site.mat2$site)
bdiv2
```
among all sites *only single-celled organisms*
```{r}
prey.sp.site.mat2b <- smdat %>%
  dplyr::select(site, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="site", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv2b <- as.matrix(betadiver(prey.sp.site.mat2b, method="j")) 
colnames(bdiv2b) <-c(prey.sp.site.mat2b$site); rownames(bdiv2b) <- c(prey.sp.site.mat2b$site)
bdiv2b
```


Then, between estuaries *all taxa*
```{r}
prey.sp.est.mat <- alldat %>%
  dplyr::select(estuary, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="estuary", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv3 <- as.matrix(betadiver(prey.sp.est.mat, method="j")) 
colnames(bdiv3) <-c(prey.sp.est.mat$estuary); rownames(bdiv3) <- c(prey.sp.est.mat$estuary)
bdiv3
```

*excluding single-celled organisms*
```{r}
prey.sp.est.mat2 <- dat %>%
  dplyr::select(estuary, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="estuary", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv4 <- as.matrix(betadiver(prey.sp.est.mat2, method="j")) 
colnames(bdiv4) <-c(prey.sp.est.mat2$estuary); rownames(bdiv4) <- c(prey.sp.est.mat2$estuary)
bdiv4
```
*only single-celled organisms*
```{r}
prey.sp.est.mat2b <- smdat %>%
  dplyr::select(estuary, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="estuary", names_from="taxon",values_from="presence", values_fill=0)

## jaccard
bdiv4b <- as.matrix(betadiver(prey.sp.est.mat2b, method="j")) 
colnames(bdiv4b) <-c(prey.sp.est.mat2b$estuary); rownames(bdiv4b) <- c(prey.sp.est.mat2b$estuary)
bdiv4b
```


save the all-taxa matrices
```{r}
write_csv(as.data.frame(bdiv, row.names=rownames(bdiv)) %>% rownames_to_column("x"), here('data','results','betadiver_runs1-2_lerayXT_presence_allTaxa.csv'))

write_csv(as.data.frame(bdiv3, row.names=rownames(bdiv3)) %>% rownames_to_column("x"), here('data','results','betadiver_estuaries_runs1-2_lerayXT_presence_allTaxa.csv'))
```

plot how within / between estuary differences change for larger and single-celled organisms. 
```{r}




```





### Zeta diversity

Zeta diversity is the mean number of species shared by any number of assemblages, and can differentiate the contribution of rare and common species to biodiversity. 

#### zeta decay across all sites

Zeta diversity decay measures the decreasing number of shared taxa across greater observed communities 

data.spec input: presence-absence data frame, with sites as rows and species as columns. I have 7 sites, so I'll ask for the zeta diversity across 7 orders?
```{r}
zdata <- t(prey.sp.site.mat %>% column_to_rownames("site"))

zout <- Zeta.decline.ex(zdata, orders = 1:7,plot=TRUE)
```

So, I read an asymptote after $\zeta_5$ - $\zeta_6$. I think this means that once I reach 5-6 sites, the average number of species shared between sites gets really small? 

Based on AIC, the exponential regression is the better fit
```{r}
zout$aic
```


```{r}
zeta.val.df <- data.frame(zorder = zout$zeta.order,
           zval = zout$zeta.val,
           zval.sd = zout$zeta.val.sd) 

zp1 <- zeta.val.df %>%
  ggplot(aes(x=zorder,y=zval, ymin=zval-zval.sd, ymax=zval+zval.sd)) +
  geom_point() + geom_path() + 
  geom_ribbon(alpha=0, color="gray60") + 
  labs(x="Zeta order (sites)", y="Zeta diversity") + theme_bw()
zp1
```


```{r}
zp2 <- zeta.val.df %>%
  ggplot(aes(x=zorder,y=zval)) + 
  geom_point() + 
  stat_smooth(method='lm', se=FALSE, color="black", size=0.5) +
  scale_y_continuous(trans='log', breaks = c(0.005,0.020,0.100,0.500,2.00)) + 
  labs(x="Zeta order (sites)", y="Zeta diversity (log)") + theme_bw()
```

```{r}
png(here('figs','figS_zeta-decay_sites_allTaxa.png'),res=300,height=1200,width=1700)
plot_grid(zp1,zp2)
dev.off()
```


#### zeta decay within an estuary (all taxa)

Within an estuary, what is the zeta decay (1) among samples, and (2) among sites?
```{r}
zin.df <- alldat %>%
  dplyr::select(estuary,site,crab_id,taxon) %>% distinct() %>%
  mutate(presence=1)
  

zdecay_samples <- list()
zdecay_sites <- list()
i <- 1

for(e in unique(zin.df$estuary)){
  ## samples
  tmp.mat.samples <- zin.df %>% filter(estuary==e) %>% dplyr::select(crab_id,taxon,presence) %>%
    pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0) %>%
    column_to_rownames("crab_id")
  zout <- Zeta.decline.ex(tmp.mat.samples, orders = 1:dim(tmp.mat.samples)[1],plot=TRUE)
  
  ## sites
  tmp.mat.sites <- zin.df %>% filter(estuary==e) %>% dplyr::select(site,taxon,presence) %>%
    distinct() %>%
    pivot_wider(id_cols="site", names_from="taxon",values_from="presence", values_fill=0) %>%
    column_to_rownames("site")
  zout.sites <- Zeta.decline.ex(tmp.mat.sites, orders = 1:dim(tmp.mat.sites)[1],plot=TRUE)
  
  zdecay_samples[[i]] <- zout
  zdecay_sites[[i]] <- zout.sites
  i <- i+1
}
names(zdecay_samples) <- unique(zin.df$estuary)

```

plot zeta diversity and zeta ratio x order for each estuary, from samples
```{r}
zeta.samples.df <- data.frame(estuary=as.character(),
                      zorder=as.numeric(),
                      zval=as.numeric(),
                      zval.sd=as.numeric(),
                      zratio=as.numeric())
for(i in seq(1,length(zdecay_samples))){
  tmpz <- zdecay_samples[[i]]
  j <- length(tmpz$combinations[which(tmpz$combinations > 1)])
  zeta.samples.df %<>% bind_rows(data.frame(zorder=tmpz$zeta.order[1:j],
                                    zval=tmpz$zeta.val[1:j],
                                    zval.sd=tmpz$zeta.val.sd[1:j],
                                    zratio=tmpz$ratio[1:j],
                                    estuary=rep(names(zdecay_samples)[i],j)))
}
```
```{r}
zplot1 <- zeta.samples.df %>%
  dplyr::select(estuary,zorder,zratio,zval) %>%
  pivot_longer(cols=c("zratio","zval"),names_to="metric",values_to="val") %>%
  mutate(estuary=factor(estuary,levels=estuary_colors$estuary),
         metric=ifelse(metric=="zval","Zeta Diversity","Ratio of Zeta Diversity")) %>%
ggplot(aes(x=zorder,y=val,color=estuary,pch=estuary)) +
  geom_point() +
  geom_path() + 
  facet_wrap(~metric,scales="free_y") +
  scale_color_manual(values=estuary_colors$key, name="estuary") +
  labs(x="Zeta order (samples)",y="") +
  theme_bw()
zplot1
```

#### zeta decay within sites (all taxa)

```{r}
zdecay_samples_bysite <- list()
i <- 1

for(s in unique(zin.df$site)){
  ## samples
  tmp.mat.samples <- zin.df %>% filter(site==s) %>% dplyr::select(crab_id,taxon,presence) %>%
    pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0) %>%
    column_to_rownames("crab_id")
  zout <- Zeta.decline.ex(tmp.mat.samples, orders = 1:dim(tmp.mat.samples)[1],plot=TRUE)
  
  zdecay_samples_bysite[[i]] <- zout
  i <- i+1
}
names(zdecay_samples_bysite) <- unique(zin.df$site)

```

plot zeta diversity and zeta ratio x order for each site
```{r}
zeta.samples.bysite.df <- data.frame(site=as.character(),
                      zorder=as.numeric(),
                      zval=as.numeric(),
                      zval.sd=as.numeric(),
                      zratio=as.numeric())
for(i in seq(1,length(zdecay_samples_bysite))){
  tmpz <- zdecay_samples_bysite[[i]]
  j <- length(tmpz$combinations[which(tmpz$combinations > 1)])
  zeta.samples.bysite.df %<>% bind_rows(data.frame(zorder=tmpz$zeta.order[1:j],
                                    zval=tmpz$zeta.val[1:j],
                                    zval.sd=tmpz$zeta.val.sd[1:j],
                                    zratio=tmpz$ratio[1:j],
                                    site=rep(names(zdecay_samples_bysite)[i],j)))
}
```
```{r}
zplot2 <- zeta.samples.bysite.df %>%
  filter(site!="PBNERR" & site!="SAMT") %>%
  left_join(site_names) %>%
  dplyr::select(site,site_name,zorder,zratio,zval) %>%
  pivot_longer(cols=c("zratio","zval"),names_to="metric",values_to="val") %>%
  mutate(metric=ifelse(metric=="zval","Zeta Diversity","Ratio of Zeta Diversity")) %>%
ggplot(aes(x=zorder,y=val,color=fct_relevel(site_name, c("Kayak Pt 1","Kayak Pt 2","March Pt","Samish Isl","Larrabee")))) +
  geom_point() +
  geom_path() + 
  facet_wrap(~metric,scales="free_y") +
  scale_color_manual(values=c("darkgreen","chartreuse3","#1f78b4","chocolate4","chocolate2"),name="site") +
  labs(x="Zeta order (samples)",y="") +
  theme_bw()
zplot2
```

faster decay = greater community overlap = indicative of lower overall diversity

```{r}
png(here('figs','figS_zeta_bysamples.png'),res=300,height=2000,width=2000)
plot_grid(zplot1 + theme(axis.title.x=element_blank(),
                         plot.margin = margin(5.5,5.5,5.5,10.5)),zplot2 + theme(plot.margin = margin(5.5,25.5,5.5,5.5)),
          nrow=2, labels=c("(a)","(b)"))
dev.off()
```
```{r}
saveRDS(zdecay_samples, here('data','results','zeta.decline.ex_bysample_forEstuaries.rds'))
saveRDS(zdecay_samples_bysite, here('data','results','zeta.decline.ex_bysample_forSites.rds'))
```



#### zeta decay within an estuary (excluding single-celled organisms)

Within an estuary, what is the zeta decay (1) among samples, and (2) among sites?
```{r}
zin.df2 <- dat %>%
  dplyr::select(estuary,site,crab_id,taxon) %>% distinct() %>%
  mutate(presence=1)
  

zdecay_samples2 <- list()
i <- 1

for(e in unique(zin.df2$estuary)){
  ## samples
  tmp.mat.samples <- zin.df2 %>% filter(estuary==e) %>% dplyr::select(crab_id,taxon,presence) %>%
    pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0) %>%
    column_to_rownames("crab_id")
  zout <- Zeta.decline.ex(tmp.mat.samples, orders = 1:dim(tmp.mat.samples)[1],plot=TRUE)
  
  zdecay_samples2[[i]] <- zout
  i <- i+1
}
names(zdecay_samples2) <- unique(zin.df2$estuary)

```

plot zeta diversity and zeta ratio x order for each estuary, from samples
```{r}
zeta.samples.df2 <- data.frame(estuary=as.character(),
                      zorder=as.numeric(),
                      zval=as.numeric(),
                      zval.sd=as.numeric(),
                      zratio=as.numeric())
for(i in seq(1,length(zdecay_samples2))){
  tmpz <- zdecay_samples2[[i]]
  j <- length(tmpz$combinations[which(tmpz$combinations > 1)])
  zeta.samples.df2 %<>% bind_rows(data.frame(zorder=tmpz$zeta.order[1:j],
                                    zval=tmpz$zeta.val[1:j],
                                    zval.sd=tmpz$zeta.val.sd[1:j],
                                    zratio=tmpz$ratio[1:j],
                                    estuary=rep(names(zdecay_samples)[i],j)))
}
```
```{r}
zplot3 <- zeta.samples.df2 %>%
  dplyr::select(estuary,zorder,zratio,zval) %>%
  pivot_longer(cols=c("zratio","zval"),names_to="metric",values_to="val") %>%
  mutate(estuary=factor(estuary,levels=estuary_colors$estuary),
         metric=ifelse(metric=="zval","Zeta Diversity","Ratio of Zeta Diversity")) %>%
ggplot(aes(x=zorder,y=val,color=estuary,pch=estuary)) +
  geom_point() +
  geom_path() + 
  facet_wrap(~metric,scales="free_y") +
  scale_color_manual(values=estuary_colors$key, name="estuary") +
  labs(x="Zeta order (samples)",y="", title="multi-cellular prey only") +
  theme_bw()
zplot3
```


#### zeta decay within sites (excluding single-celled organisms)

```{r}
zdecay_samples_bysite2 <- list()
i <- 1

for(s in unique(zin.df2$site)){
  if(s != "SAMT"){
  ## samples
  tmp.mat.samples <- zin.df2 %>% filter(site==s) %>% dplyr::select(crab_id,taxon,presence) %>%
    pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0) %>%
    column_to_rownames("crab_id")
  zout <- Zeta.decline.ex(tmp.mat.samples, orders = 1:dim(tmp.mat.samples)[1],plot=TRUE)
  
  zdecay_samples_bysite2[[i]] <- zout
  i <- i+1
  }
}
names(zdecay_samples_bysite2) <- unique(zin.df2$site)[-which(unique(zin.df2$site) =="SAMT")]

```

plot zeta diversity and zeta ratio x order for each site
```{r}
zeta.samples.bysite.df2 <- data.frame(site=as.character(),
                      zorder=as.numeric(),
                      zval=as.numeric(),
                      zval.sd=as.numeric(),
                      zratio=as.numeric())
for(i in seq(1,length(zdecay_samples_bysite2))){
  tmpz <- zdecay_samples_bysite2[[i]]
  j <- length(tmpz$combinations[which(tmpz$combinations > 1)])
  zeta.samples.bysite.df2 %<>% bind_rows(data.frame(zorder=tmpz$zeta.order[1:j],
                                    zval=tmpz$zeta.val[1:j],
                                    zval.sd=tmpz$zeta.val.sd[1:j],
                                    zratio=tmpz$ratio[1:j],
                                    site=rep(names(zdecay_samples_bysite2)[i],j)))
}
```
```{r}
zplot4 <- zeta.samples.bysite.df2 %>%
  filter(site!="PBNERR" & site!="SAMT") %>%
  left_join(site_names) %>%
  dplyr::select(site,site_name,zorder,zratio,zval) %>%
  pivot_longer(cols=c("zratio","zval"),names_to="metric",values_to="val") %>%
  mutate(metric=ifelse(metric=="zval","Zeta Diversity","Ratio of Zeta Diversity")) %>%
ggplot(aes(x=zorder,y=val,color=fct_relevel(site_name, c("Kayak Pt 1","Kayak Pt 2","March Pt","Samish Isl","Larrabee")))) +
  geom_point() +
  geom_path() + 
  facet_wrap(~metric,scales="free_y") +
  scale_color_manual(values=c("darkgreen","chartreuse3","#1f78b4","chocolate4","chocolate2"),name="site") +
  labs(x="Zeta order (samples)",y="") +
  theme_bw()
zplot4
```

faster decay = greater community overlap = indicative of lower overall diversity

```{r}
png(here('figs','figS_zeta_bysamples_onlyLgTax.png'),res=300,height=2000,width=2000)
plot_grid(zplot3 + theme(axis.title.x=element_blank(),
                         plot.margin = margin(5.5,5.5,5.5,10.5)),zplot4 + theme(plot.margin = margin(5.5,25.5,5.5,5.5)),
          nrow=2, labels=c("(a)","(b)"))
dev.off()
```
```{r}
saveRDS(zdecay_samples2, here('data','results','zeta.decline.ex_bysample_forEstuaries_LgTaxOnly.rds'))
saveRDS(zdecay_samples_bysite2, here('data','results','zeta.decline.ex_bysample_forSites_LgTaxOnly.rds'))
```





## Ordinations: Estuary / Site Differences

### Presence / absence

#### nMDS

**is this a good option for compositional data? **

Run on a presence/absence matrix, using the Jaccard dissimilarity index.
```{r}
crab.sp.mat <- dat %>%
  dplyr::select(crab_id, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0)
crabs <- crab.sp.mat$crab_id
crab.sp.mat <- as.matrix(crab.sp.mat %>% dplyr::select(-crab_id))
rownames(crab.sp.mat) <- crabs
```
```{r eval=FALSE, echo=TRUE}
z <- metaMDS(comm=crab.sp.mat, autotransform=FALSE,distance="jaccard",k=2,maxit=1000,try=40,trymax=100)
```
```{r include=FALSE}
z <- metaMDS(comm=crab.sp.mat, autotransform=FALSE,distance="jaccard",k=2,maxit=1000,try=40,trymax=100)
```
```{r}
message('stress: ', round(z$stress,3))
```
<br>
Run on Sept 15 2023; 

stress: 0.17

Run 47 stress 0.1697039 
... Procrustes: rmse 0.0001537278  max resid 0.0009080073 
... Similar to previous best
*** Best solution repeated 1 times



```{r}
#extract NMDS scores (x and y coordinates)
z.scores = as.data.frame(scores(z)$sites)
#add columns to data frame 
z.scores$Sample = rownames(z.scores)
z.scores <- left_join(z.scores,dat %>% dplyr::select(crab_id,site,estuary),by=c("Sample"="crab_id"))
z.scores <- z.scores %>% distinct() %>%
  left_join(site_names)

#make hulls
hull.data.site <- bind_rows((a <- z.scores[z.scores$site_name == "Kayak Pt 1", ][chull(z.scores[z.scores$site_name == 
    "Kayak Pt 1", c("NMDS1", "NMDS2")]), ]),
    (b <- z.scores[z.scores$site_name == "Kayak Pt 2", ][chull(z.scores[z.scores$site_name == 
    "Kayak Pt 2", c("NMDS1", "NMDS2")]), ]),  # hull values for grp B
    (c <- z.scores[z.scores$site_name == "March Pt", ][chull(z.scores[z.scores$site_name == 
    "March Pt", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "NERR", ][chull(z.scores[z.scores$site_name == 
    "NERR", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Samish Isl", ][chull(z.scores[z.scores$site_name == 
    "Samish Isl", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Samish Bay", ][chull(z.scores[z.scores$site_name == 
    "Samish Bay", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Larrabee", ][chull(z.scores[z.scores$site_name == 
    "Larrabee", c("NMDS1", "NMDS2")]), ]))

hull.data.estuary <- bind_rows((a <- z.scores[z.scores$estuary == "Port Susan Bay", ][chull(z.scores[z.scores$estuary == 
    "Port Susan Bay", c("NMDS1", "NMDS2")]), ]),
    (b <- z.scores[z.scores$estuary == "Samish Bay", ][chull(z.scores[z.scores$estuary == 
    "Samish Bay", c("NMDS1", "NMDS2")]), ]),  # hull values for grp B
    (c <- z.scores[z.scores$estuary == "Padilla Bay", ][chull(z.scores[z.scores$estuary == 
    "Padilla Bay", c("NMDS1", "NMDS2")]), ]))
```
```{r fig.height=4, fig.width=5}
#plot estuary
my.nmds <- ggplot(z.scores, aes(x = NMDS1, y = NMDS2)) + 
  geom_point(size = 2, aes(col = factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay")))) + 
  geom_polygon(data=hull.data.estuary,aes(x=NMDS1,y=NMDS2,
                                          fill=factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay")),
                                          group=factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay"))),alpha=0.30) + # add the convex hulls
  
  labs(x = "NMDS1", colour = "estuary", y = "NMDS2")  +
  scale_color_manual(values=estuary_colors$key) +
  scale_fill_manual(values=estuary_colors$key, name="estuary") + guides(fill="none") +
  theme_bw() + theme(axis.text=element_blank())
my.nmds
```
<br>

<br>

Save plot.
```{r echo=TRUE}
png(here('figs','fig3_nmds.png'), res=200, height=900,width=900)
my.nmds
dev.off()
```
<br>
<br>
```{r fig.height=4, fig.width=5}
#plot sites
my.nmds2 <- ggplot(z.scores, aes(x = NMDS1, y = NMDS2)) + 
  geom_point(size = 2, aes(col = site_name)) + 
  geom_polygon(data=hull.data.site,aes(x=NMDS1,y=NMDS2,
                                          fill=site_name,
                                          group=site),alpha=0.30) + # add the convex hulls
  
  labs(x = "NMDS1", colour = "site", y = "NMDS2")  + guides(fill="none") +
  theme_bw() + theme(axis.text=element_blank())
my.nmds2
```
Save plot.
```{r echo=TRUE}
png(here('figs','fig3_nmds-site.png'), res=200, height=900,width=900)
my.nmds2
dev.off()
```
<br>
<br>


### eDNA Index

#### nMDS 

run by calculating the distance between all technical replicates
```{r eval=FALSE, echo=TRUE}
z <- metaMDS(comm=t(index.mat), autotransform=FALSE,distance="bray",k=2,maxit=1000,try=40,trymax=100)
```

```{r}
message('stress: ', round(z$stress,3))
```
9/15/2023
Run 100 stress 0.0001711876 
... Procrustes: rmse 0.06489789  max resid 0.5287497 
*** Best solution was not repeated -- monoMDS stopping criteria:
   100: scale factor of the gradient < sfgrmin
Warning: stress is (nearly) zero: you may have insufficient data

#### PERMANOVA

First by **estuary**

```{r}
perm.index.indat <- index.df %>% 
  dplyr::select(c(1,5,6:118))
```

```{r}
site.index2.perm <- adonis2(perm.index.indat[,4:dim(perm.index.indat)[2]]~estuary, data=perm.index.indat, method="bray")
site.index2.perm
```

The permanova is highly significant

Is the significance due to the differences in centroids or due to the differences in dispersion of samples in principal coordinate space of dissimilarity (PCoA)?
```{r}
index.indat.dist <- vegdist(x=perm.index.indat[,4:dim(perm.index.indat)[2]],method="bray")
estuary.index.perm.betadisper <- betadisper(d=index.indat.dist, group=perm.index.indat$estuary)

anova(betadisper(d=index.indat.dist, perm.index.indat$estuary))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```
Significant result from PERMDISP: groups differ in dispersion 

And here's the PCoA from the PERMDISP.
```{r}
myplot <- gg_ordiplot_customLines(ord=estuary.index.perm.betadisper, groups=perm.index.indat$estuary, 
                                  groups_plotdiff="color,shape",hull=FALSE, ellipse=FALSE, label=FALSE, spiders=TRUE, 
                                  pt.size=2, line.alpha=0.5)
myplot$plot + 
  scale_shape_manual(values=c(15,17, 16), name="Estuary") +
  scale_color_manual(values=estuary_colors$key[c(2,1,3)], name="Estuary") +
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
```
```{r echo=TRUE}
png(here('figs','fig3_PCoA-estuary.png'), res=200, height=900,width=900)
myplot$plot + 
  scale_shape_manual(values=c(15,17, 16), name="Estuary") +
  scale_color_manual(values=estuary_colors$key[c(2,1,3)], name="Estuary") +
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
dev.off()
```
<br>
<br>

Then by **site**

```{r}
site.index2.perm <- adonis2(perm.index.indat[,4:dim(perm.index.indat)[2]]~site_name, data=perm.index.indat, method="bray")
site.index2.perm
```

The permanova is highly significant

Is the significance due to the differences in centroids or due to the differences in dispersion of samples in principal coordinate space of dissimilarity (PCoA)?
```{r}
index.indat.dist <- vegdist(x=perm.index.indat[,4:dim(perm.index.indat)[2]],method="bray")
site.index.perm.betadisper <- betadisper(d=index.indat.dist, group=perm.index.indat$site_name)

anova(betadisper(d=index.indat.dist, perm.index.indat$site_name))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```
Significant result from PERMDISP: groups differ in dispersion 

Save the PCoA from the PERMDISP.
```{r}
myplot2 <- gg_ordiplot_customLines(ord=site.index.perm.betadisper, groups=perm.index.indat$site_name, 
                                  groups_plotdiff="color,shape",hull=FALSE, ellipse=FALSE, label=FALSE, spiders=TRUE, 
                                  pt.size=2, line.alpha=0.5)

png(here('figs','supplement_PCoA-site.png'), res=200, height=900,width=900)
myplot2$plot + 
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
dev.off()
```

### eDNA Index x Size class

Before learning correlations for compositional data, see what happens when we impose size classes based somewhat off of instar stages: CW < 6mm, 6-8mm, 8-11.5mm, >11.5mm

Add CWs and size class
```{r}
index.df.cw <- left_join(index.df,lab_metadat %>% dplyr::select(sample_id,CW_mm), by=c("crab_id"="sample_id")) %>%
  mutate(size_class=ifelse(CW_mm < 6, "j1",
                           ifelse(CW_mm < 8, "j2",
                                  ifelse(CW_mm < 11.5, "j3","j4"))))

permCW.index.indat <- index.df.cw %>% 
  dplyr::select(c(1,120,7:118))
```

Run the PERMANOVA / PERMDISP
```{r}
cw.index.perm <- adonis2(permCW.index.indat[,3:dim(permCW.index.indat)[2]]~size_class, data=permCW.index.indat, method="bray")
cw.index.perm
```

```{r}
indexCW.indat.dist <- vegdist(x=permCW.index.indat[,3:dim(permCW.index.indat)[2]],method="bray")

anova(betadisper(d=indexCW.indat.dist, permCW.index.indat$size_class))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```



## Prey identity

### FO%

Calculate frequency of occurrence of each taxon.
```{r}
total.crab <- length(unique(dat$crab_id))

plotdat.fo <- dat %>% mutate(plot.group=ifelse(is.na(phylum), class, phylum)) %>%
  group_by(taxon,plot.group) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  arrange(plot.group)

## detritus groups
fo.sm <- smdat %>% mutate(plot.group="single-celled organisms") %>%
  filter(crab_id %in% dat$crab_id) %>%
  group_by(taxon,plot.group) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  arrange(plot.group)

plotdat.fo %<>% bind_rows(fo.sm)

plotdat.fo %<>% left_join(data.frame(plot.group=unique(plotdat.fo$plot.group),
                       ngroup=order(unique(plotdat.fo$plot.group))),by="plot.group") %>%
  unite(plot.group, ngroup, col="plot.group.num",sep="-", remove=FALSE)
```

The groups with the highest median frequency of occurrence:
```{r}
fo.hi <- plotdat.fo %>% group_by(plot.group) %>% 
  summarise(med.fo=median(pcrab)) %>%
  slice_max(order_by=med.fo,n=3, with_ties=FALSE) %>%
  left_join(dplyr::select(plotdat.fo,plot.group,taxon) %>% distinct(),by='plot.group') %>%
  left_join(bind_rows(dat,smdat), by='taxon') %>%
  filter(crab_id %in% dat$crab_id) %>%
  
  mutate(plot.subgroup=ifelse(plot.group=="single-celled organisms" & is.na(phylum), class,
                              ifelse(plot.group=="single-celled organisms" & !is.na(phylum),phylum,
                                     ifelse(plot.group=="Arthropoda",order,
                                            ifelse(plot.group=="Mollusca",order,phylum))))) %>%

  group_by(taxon,plot.group,plot.subgroup) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  mutate(outlier=ifelse(pcrab > 50, "M. acherusicum", NA)) %>%
  arrange(plot.subgroup)
```

Identify the outliers, and the groups that will be further investigated
```{r}
findoutlier <- function(x) {
  return(x < quantile(x, .25) - 4*IQR(x) | x > quantile(x, .75) + 4*IQR(x))
}

plotdat.fo <- plotdat.fo %>%
        group_by(plot.group) %>%
        mutate(outlier = ifelse(findoutlier(pcrab), taxon, NA)) %>%
  mutate(subgroup=ifelse(plot.group %in% fo.hi$plot.group, "yes","no"))
```

Plot
```{r}
plot2a <- 
  ggplot(plotdat.fo, aes(x=plot.group,y=pcrab, fill=subgroup)) +
  geom_boxplot() + geom_hline(aes(yintercept=-0.1)) +
  geom_text_repel(aes(label=outlier), na.rm=TRUE,min.segment.length = 0.01,nudge_y=0.5, size=3) +
    scale_y_continuous(trans=squish_trans(from=40,to=70,factor=10),
                       breaks=c(0,10,20,30,40,70,75)) +
  scale_fill_manual(values=c("white","yellow")) + guides(fill="none") +
  theme_bw() + labs(x="Prey Group", y="Freq. of occurrence (FO%)") + theme(axis.text.x=element_text(angle=45, hjust=1,size=12,vjust=1.05),
                     legend.text=element_text(size=12),
                     axis.title=element_text(size=13))
plot2a
```

```{r}
plot2b <-  fo.hi %>%
  mutate(plot.group=ifelse(plot.group=="single-celled organisms", "single-celled\norganisms", plot.group)) %>%
  ggplot(aes(x=plot.subgroup,y=pcrab, fill=plot.group)) +
  geom_boxplot() + geom_hline(aes(yintercept=-0.1)) +
  facet_wrap(~plot.group, scales="free_x") +
  scale_y_continuous(trans=squish_trans(from=40,to=70,factor=10),
                       breaks=c(0,10,20,30,40,70,75)) +
  geom_text_repel(aes(label=outlier), na.rm=TRUE,min.segment.length = 0.01,nudge_y=0.5, size=3) +
  theme_bw() + labs(x="Prey Subgroup", y="Freq. of occurrence (FO%)") + theme(axis.text.x=element_text(angle=60, hjust=1,size=12),
                     legend.text=element_text(size=12),
                     axis.title=element_text(size=13),
                     strip.background=element_blank(), strip.text=element_text(size=12, vjust=0),
                     panel.spacing.x=unit(0, "lines"), legend.position="none")
plot2b
```


save together
```{r}
png(here('figs','paper_fig2_FO.png'), res=300, height=2800,width=2000)
plot_grid(plot2a,plot2b,labels=c("(a)","(b)"),nrow=2)
dev.off()
```

save frequencies of occurrence into a dataframe
```{r}
write_csv(plotdat.fo, here('data','results','freqoccur_runs1-2_lerayxt_presence.csv'))
```

How many prey were only identified in 1 crab?
```{r}
dim(plotdat.fo %>% filter(ncrab==1))[1]
dim(plotdat.fo %>% filter(ncrab==2))[2]
```
```{r}
fo.hi %>% group_by(plot.group, plot.subgroup) %>% summarise(medFO=median(pcrab)) %>% filter(plot.group %in% c("Arthropoda","Mollusca","single-celled organisms"))
```


### rare prey

Which plot groups have a high proportion of rare taxa? (rare = 1-2 crab)
```{r}
raretax <- plotdat.fo %>% group_by(plot.group) %>% mutate(grouptot=length(unique(taxon))) %>%
  filter(ncrab==1) %>% ungroup() %>%
  group_by(plot.group,grouptot,ncrab) %>% summarise(nrare=length(unique(taxon)),
                                           prare=nrare/grouptot) %>% distinct()
```


### eDNA index of common prey

What is the eDNA index across sites for the 10 most common prey items?
```{r}
fo.hi.sub <- (fo.hi %>% arrange(-pcrab))[1:10,]
```



### common prey across all sites

Note that I removed March Point data, which comes from only two crab. With the March Point data, there was only one common species across all sites. 

```{r message=FALSE}
# get total crab per site & sites to include in data set.
crab.per.site <- dat %>% group_by(Site) %>% summarise(total_crab=length(unique(crab_id))) %>% ungroup()
my.sites <- crab.per.site$Site[which(crab.per.site$Site != "MARPT")]

# summarise proportion crab per prey ID
crab.per.prey <- dat %>% group_by(Site,taxon,phylum,class,order) %>% 
  summarise(ncrab=length(unique(crab_id))) %>%
  left_join(crab.per.site, by="Site") %>%
  mutate(pcrab=ncrab/total_crab)

ubiq.sp <- dat %>% dplyr::select(Site,taxon) %>%
  distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(names_from=Site,values_from=presence, values_fill=0) %>%
  mutate(total_sites=rowSums(across(c(seq(2,(length(my.sites)+1)))))) %>%
  filter(total_sites>=length(my.sites)) %>%
  left_join(crab.per.prey, by="taxon")
```
```{r}
ggplot(ubiq.sp %>% filter(Site %in% my.sites) %>%
  mutate(site=ifelse(Site=="KAY","Kayak Point 1",ifelse(Site=="KAY-shell", "Kayak Point 2",
                                                        ifelse(Site=="LAR","Larrabee",
                                                               ifelse(Site=="MARPT","March Point",NA))))), aes(x=pcrab,y=taxon,fill=Site)) +
  geom_col(position="dodge") +
  labs(x="Prop Crab at Site",y="Species",subtitle="Ubiquitous species") +
  scale_fill_manual(values=c("cyan4","darkblue","coral3")) + theme_bw()
```
<br>

### common prey unique to sites

```{r}
out <- crab.per.prey %>% split(f=~Site)


keep.sp <- map(seq_along(out), ~ anti_join(out[[.x]], 
          bind_rows(out[-.x]), by = 'taxon') %>% 
             pull(taxon))


for(i in seq(1,length(out))){
  tmp.out <- out[[i]]
  tmp.keep <- keep.sp[[i]]
  if(i==1){
  new.out <- tmp.out %>% filter(taxon %in% tmp.keep)
  } else{new.out <- new.out %>% bind_rows(tmp.out %>% filter(taxon %in% tmp.keep))}
}

new.out.filter <- new.out %>% filter(ncrab>1)
new.out.filter
```
<br>

Diving into the Arthropods...
```{r}
dat %>% filter(phylum=="Arthropoda") %>%
  dplyr::select(taxon, phylum,class,order,genus,species,Site) %>%
  distinct() %>%
  left_join(crab.per.prey) %>%
  ggplot(aes(x=order,y=pcrab, fill=Site)) +
  geom_col(position="dodge") +
  facet_wrap(~class, scales="free_x", ncol=4) + theme(panel.spacing.x=unit(0.1, "lines"),
                                                      strip.background = element_blank(),
                                                      strip.placement = "outside")
```
<br>
<br>
