---
title: "Dungeness Diet Results 1"
subtitle: "run date: `r format(Sys.time(), '%B %d, %Y')`"
author: "M Fisher"
date: 'written 2022-12-03'
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description


Summary document to produce major results information for text, tables, figures. Includes:

1. Sample sizes: 

  - How many crab were sequenced, total and per site?
  - How many sequenced crab had putative prey in their stomach contents?

2. Prey diversity metrics:

  - alpha diversity - species and taxa, overall and per site
  - alpha diversity x stomach fullness 
  - beta diversity - What was the unique number of taxa identified per per site?
  - taxonomic distinctness (vegan doc)[https://search.r-project.org/CRAN/refmans/vegan/html/taxondive.html]
  - species richness within each phyla

3. Analyze site-by-site differences  

  - Site differences in presence / absence: PERMANOVA & PERMDISP 
  - Site differences in eDNA index: PERMANOVA & PERMDISP  
  - PERMDISP as a PCoA **Main text figure**

4. Capture Prey identifications

 - Frequency of Occurrence for prey groups **Main text figure**


Each of the following is completed for unique taxa (any level allowed in the final filtered dataset), and then again for a species-only data set.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(here)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(vegan)
library(magrittr)
library(ggordiplots)
source(here('R','squash_axis.r'))
source(here('R','ggordiplot_custom.r'))
eDNAINDEX <- function(x) { #where x is a dataframe with taxa/OTUs/etc in rows, and samples in columns
  rowMax <- function(x){apply(x, MARGIN = 1, FUN = max)}
  temp <- sweep(x, MARGIN = 2, STATS = colSums(x), FUN = "/") #create proportion for each taxon/OTU within each sample
  sweep(temp, MARGIN = 1, STATS = rowMax(temp), FUN = "/")
}

# User inputs
blastdir   <- 'data/blast'
resultsdir <- 'data/results'
run.nums <- c(1,2)
marker  <- 'lerayXT'

estuary_colors <- data.frame(estuary=c("Port Susan Bay","Padilla Bay", "Samish Bay"),
                             key=c("#33a02c","#1f78b4","chocolate3"))

site_names <- data.frame(site=c("KAY","KAY-shell","MARPT","PBNERR","SAMI","SAMT","LAR"),
                         site_name=c("Kayak Pt 1","Kayak Pt 2","March Pt",
                                   "NERR","Samish Isl","Samish Bay","Larrabee"))
```
```{r data, include=FALSE}
## lab metadata (CH,CW,fullness) ##
lab_metadat <- read_csv(here('data','metadata','crab_metadat.csv'))

## seq sample id metadata ##
run_metadat_files <- c('data/DCRB_Run1_samples.csv','data/DCRB_Run2_samples.csv')
for(i in seq(1,length(run_metadat_files))){
  tmp.meta <- read_csv(here(run_metadat_files[i])) %>% mutate(MiSeqRun=run.nums[i]) %>% mutate(crab_num=as.character(crab_num), tech=as.character(tech))
  if(i==1){    metadat <- tmp.meta  } else{
    metadat <- bind_rows(metadat,tmp.meta)
  } }

## blast output, pre-filter ##
blastout <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_blast_lca_taxonomy.csv")))

blastout %<>% dplyr::select(-mi_seq_run,-sample_id) %>%
  left_join(metadat, by=c("Sample_name","Locus"="locus","MiSeqRun")) %>%
  rename(sample_id=sample_label) %>%
  mutate(crab_num=as.numeric(crab_num)) %>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## filtered taxa ##
dat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_uniqueTaxa.csv")))

dat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))

## too small taxa (aka "detritus") ##
smdat <- read_csv(here(blastdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_SmallTaxa_uniqueTaxa.csv")))  

smdat %<>%
  mutate(site=ifelse(site=="KAY" & crab_num > 18, "KAY-shell",site)) %>%
  mutate(site=ifelse(site=="CLAY","LAR",
                     ifelse(site=="SIN","SAMI",site))) %>%
  mutate(estuary=ifelse(site %in% c("KAY","KAY-shell"),"Port Susan Bay",
                                    ifelse(site %in% c("LAR","SAMT","SAMI"), "Samish Bay","Padilla Bay")))
```
<br>

# 1. Sample sizes

What are the final sample sizes for this data set?
```{r}
dat %>%
  group_by(estuary,site) %>% summarise(`Crab count`=length(unique(crab_id)))
```
<br>

What % of crab sequenced had putative prey in their stomach contents?
```{r}
pcrab_with_preytax <- blastout %>% dplyr::select(site,crab_num,estuary) %>% distinct() %>%
  mutate(crab_num=as.character(crab_num)) %>%
  left_join(dat %>% dplyr::select(crab_num,site,estuary) %>% 
              distinct() %>% mutate(with_prey="1", crab_num=as.character(crab_num))) %>%
  mutate(with_prey=ifelse(is.na(with_prey),"0",with_prey)) %>%
  group_by(estuary,site) %>%
  mutate(total_crab=length(unique(crab_num))) %>%
  group_by(estuary,site,total_crab,with_prey) %>%
  summarise(pcrab=length(unique(crab_num))/total_crab) %>% distinct() %>%
  pivot_wider(names_from=with_prey,values_from=pcrab)
pcrab_with_preytax 
```
What % of crab sequenced had putative prey **species** in their stomach contents?
```{r}
blastout %>% dplyr::select(site,crab_num,estuary) %>% distinct() %>%
  mutate(crab_num=as.character(crab_num)) %>%
  left_join(dat %>% filter(rank=="species") %>% dplyr::select(crab_num,site,estuary) %>% 
              distinct() %>% mutate(with_prey="1", crab_num=as.character(crab_num))) %>%
  mutate(with_prey=ifelse(is.na(with_prey),"0",with_prey)) %>%
  group_by(estuary,site) %>%
  mutate(total_crab=length(unique(crab_num))) %>%
  group_by(estuary,site,total_crab,with_prey) %>%
  summarise(pcrab=round(length(unique(crab_num))/total_crab, 3)) %>% distinct() %>%
  pivot_wider(names_from=with_prey,values_from=pcrab)
```


```{r}
write_csv(pcrab_with_preytax, here(resultsdir, 'lerayXT_runs1-2_percent_crab_with_prey_taxa.csv'))
```


*Important for further analysis* - do we have the same number of technical replicates per crab?
```{r}
with(bind_rows(smdat,dat) %>% group_by(crab_id) %>% summarise(ntech=length(unique(tech))), table(ntech))
with(dat %>% group_by(crab_id) %>% summarise(ntech=length(unique(tech))), table(ntech))
```



# Taxa

Allows for non-species-level IDs, as long as they are unique 

## 2. Prey Diversity

### Alpha div

What is the total number of unique taxa observed? **99** without unicellular species
```{r}
# get species-level prey for each crab
spdat <- dat %>% filter(rank=="species") %>% dplyr::select(taxon,genus,family,order,class) %>% distinct() %>%
  pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="sp.taxonomy") %>%
  rename("sp.taxon"=taxon)

# keep non-specific ids in data set only if that taxon is not represented across the entire dataset
nonspec_dat <- dat %>% filter(rank!="species") %>% dplyr::select(taxon,rank,species,genus,family,order,class) %>% distinct()

nonspec_dat %<>% pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="taxonomy") %>%
  filter(tax.level==rank) %>%
  left_join(spdat,by=c("tax.level","taxonomy"="sp.taxonomy")) %>% filter(!is.na(sp.taxon))

dat_overall <- dat %>% anti_join(nonspec_dat,by=c("rank","taxon"="taxonomy"))

rm(spdat,nonspec_dat)

length(dat_overall %>% pull(taxon) %>% unique())
```
What is the total number of unique taxa observed among unicellular algae / diatoms / rotifers / protists? **38**
```{r}
# get species-level prey for each crab
spdat <- smdat %>% filter(rank=="species") %>% dplyr::select(taxon,genus,family,order,class) %>% distinct() %>%
  pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="sp.taxonomy") %>%
  rename("sp.taxon"=taxon)

# keep non-specific ids in data set only if that taxon is not represented across the entire dataset
nonspec_dat <- smdat %>% filter(rank!="species") %>% dplyr::select(taxon,rank,species,genus,family,order,class) %>% distinct()

nonspec_dat %<>% pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="taxonomy") %>%
  filter(tax.level==rank) %>%
  left_join(spdat,by=c("tax.level","taxonomy"="sp.taxonomy")) %>% filter(!is.na(sp.taxon))

smdat_overall <- smdat %>% anti_join(nonspec_dat,by=c("rank","taxon"="taxonomy"))
rm(spdat,nonspec_dat)


length(smdat_overall %>% pull(taxon) %>% unique())
```



What is the taxonomic richness ($\alpha$ diversity) at each site? Standardized site richness is simply the  richness / total crab from site.
```{r}
spdat <- dat %>% filter(rank=="species") %>% 
  dplyr::select(site,taxon,genus,family,order,class) %>% distinct() %>%
  pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="sp.taxonomy") %>%
  rename("sp.taxon"=taxon)

nonspec_dat <- dat %>% 
  filter(rank!="species") %>% 
  dplyr::select(site,taxon,rank,species,genus,family,order,class) %>% distinct() %>%
  pivot_longer(cols=c(genus,family,order,class),names_to="tax.level",values_to="taxonomy") %>%
  filter(tax.level==rank) %>%
  left_join(spdat,by=c("site","tax.level","taxonomy"="sp.taxonomy")) %>% filter(!is.na(sp.taxon))

dat_sites <- dat %>%
  anti_join(nonspec_dat,by=c("site","rank","taxon"="taxonomy"))
rm(spdat,nonspec_dat)

richness.site <- dat_sites %>% group_by(site) %>% summarise(richness=length(unique(taxon)),
                                                            ncrab=length(unique(crab_id))) %>%
  mutate(std.richness=richness/ncrab)
richness.site
```
<br>

what is the taxonomic richness ($\alpha$ diversity) per crab?
```{r message=FALSE}
richness.crab <- dat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>%
  left_join(estuary_colors,by="estuary")

paste0("taxonomic richness per crab ranges from ", min(richness.crab$richness), " to ", max(richness.crab$richness))
```
```{r}
plot.alpha <- richness.crab %>%
  left_join(site_names, by="site") %>%
  mutate(site_name=factor(site_name, levels=c("Kayak Pt 1","Kayak Pt 2","March Pt","NERR","Samish Isl","Samish Bay", "Larrabee"))) %>%
  ggplot(aes(x=site_name,y=richness,color=estuary)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="Unique Taxa per crab",x="Sampling Site") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  scale_color_manual(values=unique(richness.crab$key)) +
  theme_bw() + theme(axis.text.x=element_text(angle=25, hjust=1))
plot.alpha
```
<br>

Save plot.
```{r echo=TRUE}
png(here('figs','fig1_alpha-div.png'), res=300, height=1000,width=1500)
plot.alpha
dev.off()
```
<br>


How is ($\alpha$ diversity) distributed across taxonomic levels?
```{r message=FALSE}
richness.phyla.site <- dat_sites %>%
  mutate(group=ifelse(!is.na(phylum),phylum,class)) %>%
  group_by(estuary,site,group) %>%
  summarise(richness=length(unique(taxon)))  %>%
  left_join(site_names)

richness.phyla <- dat_overall %>%
  mutate(group=ifelse(!is.na(phylum),phylum,class)) %>%
  group_by(estuary,site,group) %>%
  summarise(richness=length(unique(taxon))) %>%
  arrange(richness)

porder <- unique(richness.phyla$group)

richness.phyla <- richness.phyla %>%
  mutate(estuary="Overall",site="Overall") %>%
  bind_rows(richness.phyla.site)

richness.phyla$group <- factor(richness.phyla$group, levels=porder)
```
```{r fig.height=5,fig.width=5}
richness.phyla %>%
  mutate(site_name=ifelse(is.na(site_name),"Overall",site_name)) %>%
  mutate(site_name=factor(site_name, levels=c("Kayak Pt 1","Kayak Pt 2","March Pt","NERR","Samish Isl","Samish Bay", "Larrabee","Overall"))) %>%
  mutate(estuary=factor(estuary,levels=c("Port Susan Bay","Padilla Bay", "Samish Bay", "Overall"))) %>%
  ggplot(aes(y=group,x=richness, fill=site_name)) +
  geom_col(position="dodge") +
  facet_wrap(~estuary, ncol=4) +
  labs(x="Richness",y="Taxonomic Group") +
  # scale_fill_manual(values=c("cyan4","darkblue","darkorange","coral3","gray30")) +
  theme_bw()
```
<br>
save as a table
```{r}
if(!knitting){
write_csv(richness.phyla, here('data','results','lerayXT_r1-2_tax-richness-by-site.csv'))
}
```


### Alpha div x Stomach Fullness

Consider alpha diversity with / without the 'detritus' taxa, and then 'detritus' taxa alone
```{r message=FALSE}
### stomach fullness data ###
crabdat <- lab_metadat %>%
  filter(sample_id %in% dat$crab_id) %>%
  rename(crab_id=sample_id)

### without detritus ###
richness.crab <- richness.crab %>%
  left_join(crabdat)

### with detritus ###
drichness.crab <- smdat %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>% 
  left_join(estuary_colors,by="estuary") %>%
  left_join(crabdat)

### combined ###
all.richness.crab <- bind_rows(dat,smdat) %>%
  group_by(estuary,site,crab_id) %>%
  summarise(richness=length(unique(taxon))) %>% 
  left_join(estuary_colors,by="estuary") %>%
  left_join(crabdat)
```


```{r}
p.alpha3 <- ggplot(richness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="No Detritus") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha2 <- ggplot(all.richness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="All Taxa") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()
p.alpha4 <- ggplot(drichness.crab %>% filter(!is.na(stomach_fullness)), aes(x=as.factor(stomach_fullness),y=richness)) +
  geom_boxplot() +
  geom_jitter(width=0.1) +
  labs(y="",x="", title="Only Detritus") +
  scale_y_continuous(limits=c(0,max(richness.crab$richness)+2), expand=c(0,0)) +
  theme_bw()

y.grob <- textGrob("Unique Taxa per Crab", 
                   gp=gpar(col="black"), rot=90)

x.grob <- textGrob("Stomach Fullness", 
                   gp=gpar(col="black"))


main <- plot_grid(p.alpha2,p.alpha3,p.alpha4, ncol=3)
alpha_by_fll <- plot_grid(left=y.grob,main,NULL,bottom=x.grob, rel_heights=c(1,0.05), rel_widths=c(0.05,1))
alpha_by_fll
```
<br>

Save plot with the alpha diversity x site plot
```{r echo=TRUE}
png(here('figs','figS_alpha-div_site_fullness.png'), res=200, height=1000,width=1000)
plot_grid(plot.alpha + theme(plot.margin=margin(t=0,r=0,b=10,l=20)), alpha_by_fll, nrow=2)
dev.off()
```
<br>
<br>

### Distinctness


*Taxonomic distinctness* is the averaged taxonomic distances among species or individuals in the community (Clarke & Warwick 1998, 2001). These figures display the distinctness in prey assemblages between crab at the same site (with presence/absence data, both distinctness and diversity reduce to the same index, $\Delta^{+}$, Clarke & Warwick (1998)). 

Note that this metric can only be run for those species which have their full taxonomy available:
```{r}
tmpn <- length(unique(dat %>% filter(!is.na(phylum) & !is.na(class) & !is.na(order) & !is.na(family) & !is.na(genus)) %>% pull(taxon)))
message('taxonomic distinctness calculated for ', tmpn, ' species; this represents ', round(tmpn/length(unique(dat$taxon)),2)*100, '% of all species in the dataset.')
```
<br>

$\Delta^(+)$ per site:
```{r message=FALSE, warning=FALSE}
# taxonomic distances
sp.taxonomy <- dat %>%
  dplyr::select(phylum,class,order,family,genus,species) %>% distinct() %>%
  rename("Phylum"=phylum, "Class"=class,"Order"=order,"Family"=family,"Genus"=genus) %>%
  filter((!is.na(Phylum) & !is.na(Class) & !is.na(Order)) & !is.na(Family) & !is.na(Genus))
x <- as.data.frame(sp.taxonomy %>% dplyr::select(Genus,Family,Order,Class,Phylum))
rownames(x) <- sp.taxonomy$species
sp.dist.mat <- taxa2dist(x, varstep = FALSE, check = TRUE, labels=rownames(x))

# presence / absence matrix
prey.sp.site.df <- dat %>%
  filter(species %in% sp.taxonomy$species) %>%
  dplyr::select(Site, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="Site", names_from="taxon",values_from="presence", values_fill=0)

prey.sp.site.mat <- as.data.frame(prey.sp.site.df %>% dplyr::select(-Site))
rownames(prey.sp.site.mat) <- prey.sp.site.df$Site


# calculate distance
site.taxdist <- taxondive(comm=prey.sp.site.mat, dis=sp.dist.mat, match.force=TRUE)
site.taxdist.df <- data.frame(site=names(site.taxdist$Species),
                              nsp=site.taxdist$Species,
                              dplus=site.taxdist$Dplus,
                              dplus.sd=site.taxdist$sd.Dplus) %>%
  mutate(site=ifelse(site=="KAY","Kayak\nPoint 1",ifelse(site=="KAY-shell", "Kayak\nPoint 2",
                                                        ifelse(site=="LAR","Larrabee",
                                                               ifelse(site=="MARPT","March\nPoint",NA)))))
```
```{r fig.width=5, fig.height=4}
pd1 <- ggplot(site.taxdist.df, aes(x=site,y=dplus, col=nsp)) +
  geom_point() +
  geom_errorbar(aes(ymin=dplus-dplus.sd,ymax=dplus+dplus.sd), width=0.5) +
  scale_y_continuous(limits=c(75,100)) +
  labs(x="Site",y="Distinctness", title="Taxonomic Distinctness",subtitle="*calculated from 86% of species in data set") +
  scale_color_gradient(low = "plum2", high = "coral4", na.value = NA, name="Species\n Count") +
  theme_bw()

pd1
```

<br>

Save plot.
```{r echo=TRUE}
png(here('figs','fig2_distinctness.png'), res=200, height=600,width=800)
pd1
dev.off()
```
<br>
<br>


<br>
<br>


### eDNA Index

calculate eDNA index for all technical replicates
```{r}
crab.reads.mat <- dat %>%
  group_by(sample_id, taxon) %>% summarise(techReads=sum(nReads)) %>%
  pivot_wider(id_cols="sample_id", names_from="taxon",values_from="techReads", values_fill=0) %>%
  column_to_rownames("sample_id")

index.mat <- eDNAINDEX(x=t(crab.reads.mat))

index.df <- as.data.frame(t(index.mat)) %>%
  rownames_to_column("sample_id") %>% 
  pivot_longer(cols=2:(dim(index.mat)[1]+1), names_to="taxon",values_to="DNAindex") %>%
  left_join(dat %>% dplyr::select(sample_id,crab_id,tech,site,estuary) %>% distinct(),by="sample_id") %>%
  left_join(site_names) %>%
  pivot_wider(names_from=taxon,values_from=DNAindex)
```



## Ordinations: Estuary / Site Differences

### Presence / absence

#### nMDS

**is this a good option for compositional data? **

Run on a presence/absence matrix, using the Jaccard dissimilarity index.
```{r}
crab.sp.mat <- dat %>%
  dplyr::select(crab_id, taxon) %>% distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(id_cols="crab_id", names_from="taxon",values_from="presence", values_fill=0)
crabs <- crab.sp.mat$crab_id
crab.sp.mat <- as.matrix(crab.sp.mat %>% dplyr::select(-crab_id))
rownames(crab.sp.mat) <- crabs
```
```{r eval=FALSE, echo=TRUE}
z <- metaMDS(comm=crab.sp.mat, autotransform=FALSE,distance="jaccard",k=2,maxit=1000,try=40,trymax=100)
```
```{r include=FALSE}
z <- metaMDS(comm=crab.sp.mat, autotransform=FALSE,distance="jaccard",k=2,maxit=1000,try=40,trymax=100)
```
```{r}
message('stress: ', round(z$stress,3))
```
<br>
Run on Sept 15 2023; 

stress: 0.17

Run 47 stress 0.1697039 
... Procrustes: rmse 0.0001537278  max resid 0.0009080073 
... Similar to previous best
*** Best solution repeated 1 times



```{r}
#extract NMDS scores (x and y coordinates)
z.scores = as.data.frame(scores(z)$sites)
#add columns to data frame 
z.scores$Sample = rownames(z.scores)
z.scores <- left_join(z.scores,dat %>% dplyr::select(crab_id,site,estuary),by=c("Sample"="crab_id"))
z.scores <- z.scores %>% distinct() %>%
  left_join(site_names)

#make hulls
hull.data.site <- bind_rows((a <- z.scores[z.scores$site_name == "Kayak Pt 1", ][chull(z.scores[z.scores$site_name == 
    "Kayak Pt 1", c("NMDS1", "NMDS2")]), ]),
    (b <- z.scores[z.scores$site_name == "Kayak Pt 2", ][chull(z.scores[z.scores$site_name == 
    "Kayak Pt 2", c("NMDS1", "NMDS2")]), ]),  # hull values for grp B
    (c <- z.scores[z.scores$site_name == "March Pt", ][chull(z.scores[z.scores$site_name == 
    "March Pt", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "NERR", ][chull(z.scores[z.scores$site_name == 
    "NERR", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Samish Isl", ][chull(z.scores[z.scores$site_name == 
    "Samish Isl", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Samish Bay", ][chull(z.scores[z.scores$site_name == 
    "Samish Bay", c("NMDS1", "NMDS2")]), ]),
    (d <- z.scores[z.scores$site_name == "Larrabee", ][chull(z.scores[z.scores$site_name == 
    "Larrabee", c("NMDS1", "NMDS2")]), ]))

hull.data.estuary <- bind_rows((a <- z.scores[z.scores$estuary == "Port Susan Bay", ][chull(z.scores[z.scores$estuary == 
    "Port Susan Bay", c("NMDS1", "NMDS2")]), ]),
    (b <- z.scores[z.scores$estuary == "Samish Bay", ][chull(z.scores[z.scores$estuary == 
    "Samish Bay", c("NMDS1", "NMDS2")]), ]),  # hull values for grp B
    (c <- z.scores[z.scores$estuary == "Padilla Bay", ][chull(z.scores[z.scores$estuary == 
    "Padilla Bay", c("NMDS1", "NMDS2")]), ]))
```
```{r fig.height=4, fig.width=5}
#plot estuary
my.nmds <- ggplot(z.scores, aes(x = NMDS1, y = NMDS2)) + 
  geom_point(size = 2, aes(col = factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay")))) + 
  geom_polygon(data=hull.data.estuary,aes(x=NMDS1,y=NMDS2,
                                          fill=factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay")),
                                          group=factor(estuary, levels=c("Port Susan Bay","Padilla Bay","Samish Bay"))),alpha=0.30) + # add the convex hulls
  
  labs(x = "NMDS1", colour = "estuary", y = "NMDS2")  +
  scale_color_manual(values=estuary_colors$key) +
  scale_fill_manual(values=estuary_colors$key, name="estuary") + guides(fill="none") +
  theme_bw() + theme(axis.text=element_blank())
my.nmds
```
<br>

<br>

Save plot.
```{r echo=TRUE}
png(here('figs','fig3_nmds.png'), res=200, height=900,width=900)
my.nmds
dev.off()
```
<br>
<br>
```{r fig.height=4, fig.width=5}
#plot sites
my.nmds2 <- ggplot(z.scores, aes(x = NMDS1, y = NMDS2)) + 
  geom_point(size = 2, aes(col = site_name)) + 
  geom_polygon(data=hull.data.site,aes(x=NMDS1,y=NMDS2,
                                          fill=site_name,
                                          group=site),alpha=0.30) + # add the convex hulls
  
  labs(x = "NMDS1", colour = "site", y = "NMDS2")  + guides(fill="none") +
  theme_bw() + theme(axis.text=element_blank())
my.nmds2
```
Save plot.
```{r echo=TRUE}
png(here('figs','fig3_nmds-site.png'), res=200, height=900,width=900)
my.nmds2
dev.off()
```
<br>
<br>


### eDNA Index

#### nMDS 

run by calculating the distance between all technical replicates
```{r eval=FALSE, echo=TRUE}
z <- metaMDS(comm=t(index.mat), autotransform=FALSE,distance="bray",k=2,maxit=1000,try=40,trymax=100)
```

```{r}
message('stress: ', round(z$stress,3))
```
9/15/2023
Run 100 stress 0.0001711876 
... Procrustes: rmse 0.06489789  max resid 0.5287497 
*** Best solution was not repeated -- monoMDS stopping criteria:
   100: scale factor of the gradient < sfgrmin
Warning: stress is (nearly) zero: you may have insufficient data

#### PERMANOVA

First by **estuary**

```{r}
perm.index.indat <- index.df %>% 
  dplyr::select(c(1,5,6:118))
```

```{r}
site.index2.perm <- adonis2(perm.index.indat[,4:dim(perm.index.indat)[2]]~estuary, data=perm.index.indat, method="bray")
site.index2.perm
```

The permanova is highly significant

Is the significance due to the differences in centroids or due to the differences in dispersion of samples in principal coordinate space of dissimilarity (PCoA)?
```{r}
index.indat.dist <- vegdist(x=perm.index.indat[,4:dim(perm.index.indat)[2]],method="bray")
estuary.index.perm.betadisper <- betadisper(d=index.indat.dist, group=perm.index.indat$estuary)

anova(betadisper(d=index.indat.dist, perm.index.indat$estuary))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```
Significant result from PERMDISP: groups differ in dispersion 

And here's the PCoA from the PERMDISP.
```{r}
myplot <- gg_ordiplot_customLines(ord=estuary.index.perm.betadisper, groups=perm.index.indat$estuary, 
                                  groups_plotdiff="color,shape",hull=FALSE, ellipse=FALSE, label=FALSE, spiders=TRUE, 
                                  pt.size=2, line.alpha=0.5)
myplot$plot + 
  scale_shape_manual(values=c(15,17, 16), name="Estuary") +
  scale_color_manual(values=estuary_colors$key[c(2,1,3)], name="Estuary") +
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
```
```{r echo=TRUE}
png(here('figs','fig3_PCoA-estuary.png'), res=200, height=900,width=900)
myplot$plot + 
  scale_shape_manual(values=c(15,17, 16), name="Estuary") +
  scale_color_manual(values=estuary_colors$key[c(2,1,3)], name="Estuary") +
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
dev.off()
```
<br>
<br>

Then by **site**

```{r}
site.index2.perm <- adonis2(perm.index.indat[,4:dim(perm.index.indat)[2]]~site_name, data=perm.index.indat, method="bray")
site.index2.perm
```

The permanova is highly significant

Is the significance due to the differences in centroids or due to the differences in dispersion of samples in principal coordinate space of dissimilarity (PCoA)?
```{r}
index.indat.dist <- vegdist(x=perm.index.indat[,4:dim(perm.index.indat)[2]],method="bray")
site.index.perm.betadisper <- betadisper(d=index.indat.dist, group=perm.index.indat$site_name)

anova(betadisper(d=index.indat.dist, perm.index.indat$site_name))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```
Significant result from PERMDISP: groups differ in dispersion 

Save the PCoA from the PERMDISP.
```{r}
myplot2 <- gg_ordiplot_customLines(ord=site.index.perm.betadisper, groups=perm.index.indat$site_name, 
                                  groups_plotdiff="color,shape",hull=FALSE, ellipse=FALSE, label=FALSE, spiders=TRUE, 
                                  pt.size=2, line.alpha=0.5)

png(here('figs','supplement_PCoA-site.png'), res=200, height=900,width=900)
myplot2$plot + 
  labs(x="PCoA Dim 1",y="PCoA Dim 2") +
  theme_bw() + theme(axis.text=element_blank(),axis.ticks=element_blank())
dev.off()
```

### eDNA Index x Size class

Before learning correlations for compositional data, see what happens when we impose size classes based somewhat off of instar stages: CW < 6mm, 6-8mm, 8-11.5mm, >11.5mm

Add CWs and size class
```{r}
index.df.cw <- left_join(index.df,lab_metadat %>% dplyr::select(sample_id,CW_mm), by=c("crab_id"="sample_id")) %>%
  mutate(size_class=ifelse(CW_mm < 6, "j1",
                           ifelse(CW_mm < 8, "j2",
                                  ifelse(CW_mm < 11.5, "j3","j4"))))

permCW.index.indat <- index.df.cw %>% 
  dplyr::select(c(1,120,7:118))
```

Run the PERMANOVA / PERMDISP
```{r}
cw.index.perm <- adonis2(permCW.index.indat[,3:dim(permCW.index.indat)[2]]~size_class, data=permCW.index.indat, method="bray")
cw.index.perm
```

```{r}
indexCW.indat.dist <- vegdist(x=permCW.index.indat[,3:dim(permCW.index.indat)[2]],method="bray")

anova(betadisper(d=indexCW.indat.dist, permCW.index.indat$size_class))
# adonis2(dist(estuary.index.perm.betadisper$distances)~estuary.index.perm.betadisper$group)
```



## Correlations: Size
https://rachaellappan.github.io/16S-analysis/correlation-between-otus-with-sparcc.html
https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html

or, Kelly paper with weighted correlation network analysis https://www.nature.com/articles/s41467-019-14105-1#code-availability

In order to directly associate specific taxa to environmental variables (Fig. 3), we used sparse partial least square (sPLS) analysis from the R package mixOmics57,58. We applied the sPLS in regression mode, which will model a causal relationship between the lineages and the environmental traits, that is, PLS will predict environmental traits (e.g., temperature) from lineage abundances. This approach enabled us to identify strong correlations between certain taxa and environmental variables without taking into account the global structure of the planktonic community.


## Prey identity

### FO%

Calculate frequency of occurrence of each taxon.
```{r}
total.crab <- length(unique(dat$crab_id))

plotdat.fo <- dat %>% mutate(plot.group=ifelse(is.na(phylum), class, phylum)) %>%
  group_by(taxon,plot.group) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  arrange(plot.group)

## detritus groups
fo.sm <- smdat %>% mutate(plot.group="single-celled organisms") %>%
  filter(crab_id %in% dat$crab_id) %>%
  group_by(taxon,plot.group) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  arrange(plot.group)

plotdat.fo %<>% bind_rows(fo.sm)

plotdat.fo %<>% left_join(data.frame(plot.group=unique(plotdat.fo$plot.group),
                       ngroup=order(unique(plotdat.fo$plot.group))),by="plot.group") %>%
  unite(plot.group, ngroup, col="plot.group.num",sep="-", remove=FALSE)
```

The groups with the highest median frequency of occurrence:
```{r}
fo.hi <- plotdat.fo %>% group_by(plot.group) %>% 
  summarise(med.fo=median(pcrab)) %>%
  slice_max(order_by=med.fo,n=3, with_ties=FALSE) %>%
  left_join(dplyr::select(plotdat.fo,taxon) %>% distinct(),by='plot.group') %>%
  left_join(bind_rows(dat,smdat), by='taxon') %>%
  filter(crab_id %in% dat$crab_id) %>%
  
  mutate(plot.subgroup=ifelse(plot.group=="single-celled organisms" & is.na(phylum), class,
                              ifelse(plot.group=="single-celled organisms" & !is.na(phylum),phylum,
                                     ifelse(plot.group=="Arthropoda",order,
                                            ifelse(plot.group=="Mollusca",order,phylum))))) %>%

  group_by(taxon,plot.group,plot.subgroup) %>%
  summarise(ncrab=length(unique(crab_id))) %>%
  mutate(pcrab=(ncrab/total.crab)*100) %>%
  mutate(outlier=ifelse(pcrab > 50, "M. acherusicum", NA)) %>%
  arrange(plot.subgroup)
```

Identify the outliers, and the groups that will be further investigated
```{r}
findoutlier <- function(x) {
  return(x < quantile(x, .25) - 4*IQR(x) | x > quantile(x, .75) + 4*IQR(x))
}

plotdat.fo <- plotdat.fo %>%
        group_by(plot.group) %>%
        mutate(outlier = ifelse(findoutlier(pcrab), taxon, NA)) %>%
  mutate(subgroup=ifelse(plot.group %in% fo.hi$plot.group, "yes","no"))
```

Plot
```{r}
plot2a <- 
  ggplot(plotdat.fo, aes(x=plot.group,y=pcrab, fill=subgroup)) +
  geom_boxplot() + geom_hline(aes(yintercept=-0.1)) +
  geom_text_repel(aes(label=outlier), na.rm=TRUE,min.segment.length = 0.01,nudge_y=0.5, size=3) +
    scale_y_continuous(trans=squish_trans(from=40,to=70,factor=10),
                       breaks=c(0,10,20,30,40,70,75)) +
  scale_fill_manual(values=c("white","yellow")) + guides(fill="none") +
  theme_bw() + labs(x="Prey Group", y="Freq. of occurrence (FO%)") + theme(axis.text.x=element_text(angle=45, hjust=1,size=12,vjust=1.05),
                     legend.text=element_text(size=12),
                     axis.title=element_text(size=13))
plot2a
```

```{r}
plot2b <-  fo.hi %>%
  mutate(plot.group=ifelse(plot.group=="single-celled organisms", "single-celled\norganisms", plot.group)) %>%
  ggplot(aes(x=plot.subgroup,y=pcrab, fill=plot.group)) +
  geom_boxplot() + geom_hline(aes(yintercept=-0.1)) +
  facet_wrap(~plot.group, scales="free_x") +
  scale_y_continuous(trans=squish_trans(from=40,to=70,factor=10),
                       breaks=c(0,10,20,30,40,70,75)) +
  geom_text_repel(aes(label=outlier), na.rm=TRUE,min.segment.length = 0.01,nudge_y=0.5, size=3) +
  theme_bw() + labs(x="Prey Subgroup", y="Freq. of occurrence (FO%)") + theme(axis.text.x=element_text(angle=60, hjust=1,size=12),
                     legend.text=element_text(size=12),
                     axis.title=element_text(size=13),
                     strip.background=element_blank(), strip.text=element_text(size=12, vjust=0),
                     panel.spacing.x=unit(0, "lines"), legend.position="none")
plot2b
```


save together
```{r}
png(here('figs','paper_fig2_FO.png'), res=300, height=2800,width=2000)
plot_grid(plot2a,plot2b,labels=c("(a)","(b)"),nrow=2)
dev.off()
```





### common prey across all sites

Note that I removed March Point data, which comes from only two crab. With the March Point data, there was only one common species across all sites. 

```{r message=FALSE}
# get total crab per site & sites to include in data set.
crab.per.site <- dat %>% group_by(Site) %>% summarise(total_crab=length(unique(crab_id))) %>% ungroup()
my.sites <- crab.per.site$Site[which(crab.per.site$Site != "MARPT")]

# summarise proportion crab per prey ID
crab.per.prey <- dat %>% group_by(Site,taxon,phylum,class,order) %>% 
  summarise(ncrab=length(unique(crab_id))) %>%
  left_join(crab.per.site, by="Site") %>%
  mutate(pcrab=ncrab/total_crab)

ubiq.sp <- dat %>% dplyr::select(Site,taxon) %>%
  distinct() %>%
  mutate(presence=1) %>%
  pivot_wider(names_from=Site,values_from=presence, values_fill=0) %>%
  mutate(total_sites=rowSums(across(c(seq(2,(length(my.sites)+1)))))) %>%
  filter(total_sites>=length(my.sites)) %>%
  left_join(crab.per.prey, by="taxon")
```
```{r}
ggplot(ubiq.sp %>% filter(Site %in% my.sites) %>%
  mutate(site=ifelse(Site=="KAY","Kayak Point 1",ifelse(Site=="KAY-shell", "Kayak Point 2",
                                                        ifelse(Site=="LAR","Larrabee",
                                                               ifelse(Site=="MARPT","March Point",NA))))), aes(x=pcrab,y=taxon,fill=Site)) +
  geom_col(position="dodge") +
  labs(x="Prop Crab at Site",y="Species",subtitle="Ubiquitous species") +
  scale_fill_manual(values=c("cyan4","darkblue","coral3")) + theme_bw()
```
<br>

### common prey unique to sites

```{r}
out <- crab.per.prey %>% split(f=~Site)


keep.sp <- map(seq_along(out), ~ anti_join(out[[.x]], 
          bind_rows(out[-.x]), by = 'taxon') %>% 
             pull(taxon))


for(i in seq(1,length(out))){
  tmp.out <- out[[i]]
  tmp.keep <- keep.sp[[i]]
  if(i==1){
  new.out <- tmp.out %>% filter(taxon %in% tmp.keep)
  } else{new.out <- new.out %>% bind_rows(tmp.out %>% filter(taxon %in% tmp.keep))}
}

new.out.filter <- new.out %>% filter(ncrab>1)
new.out.filter
```
<br>

Diving into the Arthropods...
```{r}
dat %>% filter(phylum=="Arthropoda") %>%
  dplyr::select(taxon, phylum,class,order,genus,species,Site) %>%
  distinct() %>%
  left_join(crab.per.prey) %>%
  ggplot(aes(x=order,y=pcrab, fill=Site)) +
  geom_col(position="dodge") +
  facet_wrap(~class, scales="free_x", ncol=4) + theme(panel.spacing.x=unit(0.1, "lines"),
                                                      strip.background = element_blank(),
                                                      strip.placement = "outside")
```
<br>
<br>
