---
title: "cutadapt.wrapper"
author: "M Fisher, via Eily via Moncho"
date: "2022-09-30"
output: 
  html_document:
    toc: yes
---



# Description

Use [*cutadapt*](doi.org/10.14806/ej.17.1.200) to remove Nextera indices and PCR primers from DNA sequences. 
This RMarkdown doc is just a means to passing the parameters from the metadata file into a bash script.

**Inputs** for this script: 

1. file with primer sequences
2. raw, demultiplexed sequencing data in fastq.gz files

*IMPORTANT! this won't work unless you manually copy and paste in the full file path to params.txt file at bottom (code chunks = test_bash, run_cutadapt). Do that then knit.*

To install cutadapt, download the [source code](); use the `gunzip` and `tar -xvf` to unpack.
<br>
<br>


# Set up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(tidyverse)
```
<br>

User directories and inputs
```{r}
# folder with fastq files
fastq_dir <- "data/raw"

# folder for output
output_dir <- "data/cutadapt"

# folder for sequencing metadata file
meta_output_dir <- "data"

# which lane of sequencing is being analyzed?
run <- 1

# which primer was used for this lane?
run_primer <- "LerayXT"

# ?
params <- data.frame(minlength=100)
```
<br>

Where is the file with the primer sequences, and what is it called? The columns in this file should be as follows:  `primer` | `primerF` | `primerR`. The *primer* name in this file should match the name in the sequencing center metadata file -- if working across different loci, the locus name can be used in this column.  
```{r}
meta_dir <- "data"
primerfilename <- "primer_sequences.csv"
```
<br>


# Input / output files

## create sequencing metadata file

The minimum info needed for the sequencing metadata file is:

- Sample_name - The name that makes sense to you and your project (No spaces in the name)
- Locus: The name of the locus you want to use (e.g. Leray_COI)
- PrimerF: The nucleotide sequence of the forward primer - supports IUPAC characters 
- PrimerR: Ditto for the reverse primer (also in 5' -> 3' direction)
- file1: it should match exactly the output of the Miseq.
- file2: Same for the second read.
        

Read in the .csv file with the primer sequences
```{r}
primer_seqs <- read_csv(here(meta_dir, primerfilename))
```
<br>

Get all of the sequencing filenames
```{r}
## forward
ffilenames <- list.files(path = here(fastq_dir), 
                         pattern="*R1_001.fastq.gz")
## reverse
rfilenames <- list.files(path = here(fastq_dir), 
                         pattern="*R2_001.fastq.gz")
```
<br>

Grab the sample IDs and sample numbers ("name") from the filenames, arrange them in order of sample number, and combine the datasets.
```{r}
fdat <- data.frame(file1=ffilenames) %>%
  # create sample name / sample ID from file name, without file suffix
  mutate(sample_name=str_remove(file1,"_L001_R1_001.fastq.gz")) %>%
  separate(col=sample_name, into=c("sample_label","Sample_name"), sep=-2) %>%
  # clean up sample id and sample name
  mutate(sample_label=str_remove(sample_label,"_S"),
         sample_label=ifelse(str_sub(sample_label,-1)=="_",str_sub(sample_label,1,-2), sample_label),
         Sample_name=str_remove(Sample_name,"_S"),
         Sample_name=str_remove(Sample_name,"S")) %>%
  arrange(as.numeric(Sample_name))


rdat <- data.frame(file2=rfilenames) %>%
  # create sample name / sample ID from file name, without file suffix
  mutate(sample_name=str_remove(file2,"_L001_R2_001.fastq.gz")) %>%
  separate(col=sample_name, into=c("sample_label","Sample_name"), sep=-2) %>%
  # clean up sample id and sample name
  mutate(sample_label=str_remove(sample_label,"_S"),
         sample_label=ifelse(str_sub(sample_label,-1)=="_",str_sub(sample_label,1,-2), sample_label),
         Sample_name=str_remove(Sample_name,"_S"),
         Sample_name=str_remove(Sample_name,"S")) %>%
  arrange(as.numeric(Sample_name))

sequencingmetafile <- full_join(fdat,rdat,by=c("Sample_name","sample_label")) %>%
  dplyr::select(Sample_name,file1,file2)

```
<br>

Add primer information
```{r}
sequencingmetafile %<>%
  mutate(primer=run_primer) %>%
  left_join(primer_seqs, by="primer") %>%
  mutate(Locus=primer,
         PrimerF=primerF,
         PrimerR=primerR) %>%
  dplyr::select(Sample_name,file1,file2,PrimerF,PrimerR,Locus)
```
<br>

Save file
```{r}
write_csv(sequencingmetafile, here(meta_output_dir, paste0("metadata-cutadapt-input-run-", run, ".csv")))
```
<br>
<br>


## store params 

folder for output from bash script
```{r}
outputfolder <- paste0(here(output_dir))
dir.create(outputfolder)

paramsfile <- paste0(outputfolder,"/params.txt")
```
<br>

sequencing metadata file name and directory
```{r}
sequencingmetadatafile <- paste0(here(meta_output_dir, paste0("metadata-cutadapt-input-run-", run, ".csv")))
```
<br>

full file path to fastq files
```{r}
fastqfolder <- paste0(here(fastq_dir))
```
<br>

create and save the params file, which will be called in bash to run cutadapt.
```{r}
params2 <- c(fastqfolder,sequencingmetadatafile, outputfolder, params$minlength)

tibble(values = as.character(params2), names = c("fastqfolder", "sequencingmetadatafile", "outputfolder", "minlength")) %>% 
  pivot_wider(names_from = names,
              values_from = values) %>%
  write_csv(paramsfile)
```
<br>
<br>

# Cutadapt wrapper

Run the following in the terminal. This wrapper script will print some parameter values to the terminal (see below) before running the core *cutadapt* script, which is `run.cutadapt.sh`. Note that output from `run.cutadapt.sh` will be saved into a log file, instead of printed to the terminal. 
```{bash run_cutadapt, eval=FALSE}
cd ~/Documents/DungyDietAssessment

 bash scripts/1_cutadapt_wrapper.sh "C:/Users/mfisher5/Documents/DungyDietAssessment/data/cutadapt/params.txt"
```

output: 
```
/c/Users/mfisher5/Documents/DungyDietAssessment
fastqfolder,sequencingmetadatafile,outputfolder,minlength
C:/Users/mfisher5/Documents/DungyDietAssessment/data/raw,C:/Users/mfisher5/Documents/DungyDietAssessment/data/metadata-cutadapt-input-run-1.csv,C:/Users/mfisher5/Documents/DungyDietAssessment/data/cutadapt,100
awk: cmd. line:1: fatal: cannot open file ` ' for reading (No such file or directory)
fastq files will be read in from this folder:
C:/Users/mfisher5/Documents/DungyDietAssessment/data/raw
---
and trimmed fastq files will be saved into this folder:
C:/Users/mfisher5/Documents/DungyDietAssessment/data/cutadapt
---
---
```
