---
title: "Processing Blast Output: Taxonomy Clean Up"
subtitle: "run date: `r format(Sys.time(), '%B %d, %Y')`"
author: "M Fisher"
date: 'written 2022-10-12'
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description

Clean up the taxonomic information for each crab (+technical replicate). This involves:

1. Removing predator DNA

2. Removing non-target bacteria, fungal DNA


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(ggplot2)
library(ggrepel)
library(magrittr)
```

Specify user directories
```{r}
dada_dir    <- 'data/dada2'
blast_dir   <- 'data/blast'
outdir      <- 'data/blast'
```
<br>

What are the names of the processed blast file with taxonomy from script 6?
```{r}
# hashes with unique staxids
blast_file <- 'lerayXT_r1-2_sample_blast_lca_taxonomy.csv'
```
<br>

What is the location / name of the metadata file for this sequencing run?
```{r}
run_metadat_files <- c('data/DCRB_Run1_samples.csv','data/DCRB_Run2_samples.csv')
```
<br>

And the run number and the marker
```{r}
run.nums <- c(1,2)
marker <- 'lerayXT'
```
<br>

# Read in data

```{r read in asvs1, include=FALSE}
blast.dat <- read_csv(here(blast_dir, blast_file)) %>% dplyr::select(-mi_seq_run)
# remove the sample numbers at the start of the sample IDs from miseqrun 1
blast.dat <- blast.dat %>% filter(MiSeqRun==1) %>%
  separate(sample_id, c("tmp1","tmp2","tmp3","tmp4"), sep="-", remove=TRUE) %>%
  unite(col="sample_id",tmp2,tmp3,tmp4, sep="-", remove=TRUE) %>%
  mutate(sample_id=ifelse(grepl("Control",sample_id),str_remove(sample_id,"-NA"),sample_id)) %>%
  bind_rows(blast.dat %>% filter(MiSeqRun==2))

for(i in seq(1,length(run_metadat_files))){
  tmp.meta <- read_csv(here(run_metadat_files[i])) %>% mutate(MiSeqRun=run.nums[i]) %>% mutate(crab_num=as.character(crab_num),
                                                                                               tech=as.character(tech))
  if(i==1){    metadat <- tmp.meta  } else{
    metadat <- bind_rows(metadat,tmp.meta)
  }
}
```
<br>

Join the blast data to metadata information. Remove positive and negative controls (their hashes were removed in the filtering / denoising step before blasting.)
```{r}
blast.dat.filter <- blast.dat %>%
  dplyr::select(-tmp1) %>%
  left_join(metadat, by=c("Sample_name","Locus"="locus","MiSeqRun")) %>%
  filter(!(grepl("Control",sample_id))) %>%
  separate(sample_label, into=c("temp1","temp2","temp3"), sep="-", remove=FALSE) %>% unite(col="crab_id",c(temp1,temp2)) %>%
  mutate(sample_id=sample_label) %>%
  dplyr::select(-temp3,-sample_label)
```
<br>

Do all crab have three replicates present? *Yay! all of them.*
```{r}
blast.dat.filter %>% group_by(crab_id) %>% summarise(nrep=length(unique(tech))) %>% filter(nrep < 3)
```
<br>

How many crab per site are present in the blast data?
```{r}
site.totals <- blast.dat.filter %>% group_by(site) %>% summarise(total_crab=length(unique(crab_id)), total_hashes=length(unique(Hash)), total_reads=sum(nReads))
site.totals %>% dplyr::select(site,total_crab)
```
<br>

# Identify Taxonomy for Removal

To easier track how much data is being removed, I'm going to classify each hash as (1) to remove, and (2) reason for removal. then in the next section, I'll filter the data and check out the prey taxa. 

### Missing or Low Resolution Taxonomy

Remove any hashes that don't have taxonomy assigned, *OR* don't have a taxonomic rank at the order level or lower.
```{r}
blast.dat.filter %<>% mutate(missing_rm=ifelse(is.na(rank) | rank %in% c("kingdom","phylum","class"), 1,0))
```
```{r echo=FALSE, message=FALSE}
blast.dat.filter %>% group_by(site,missing_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(site.totals,by="site") %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(site,missing_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=site,y=proportion, fill=as.factor(missing_rm))) + geom_col() + facet_wrap(~metric) +
  labs(x="Site",y="Proportion") +
  scale_fill_manual(values=c("darkseagreen4","black"), name="Missing Tax") +
  theme_classic() + theme(axis.text.x=element_text(angle=45,hjust=1))
```
<br>

Let's get rid of those hashes missing taxonomy.
```{r}
blast.dat.filter %<>% filter(missing_rm==0)
```
<br>

calculate per crab metrics
```{r}
crab.totals <- blast.dat.filter %>%
  group_by(crab_id) %>% summarise(total_hashes=length(unique(Hash)), total_reads=sum(nReads))
```
<br>

### Predator DNA

Any Dungeness crab DNA will be marked as 'predator' DNA and removed. 

- Metacarcinus magister (species)
- Metacarcinus (genus)
<br>

```{r}
blast.dat.filter %<>% 
  mutate(predator_rm=ifelse(missing_rm==0 & (taxon =="Metacarcinus magister" | taxon =="Metacarcinus"), 1, 0))
```
```{r echo=FALSE, message=FALSE, fig.width=10, fig.height=5}
blast.dat.filter %>%
  group_by(site,crab_id,predator_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(site,crab_id,predator_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=crab_id,y=proportion, fill=as.factor(predator_rm))) + geom_col() + facet_grid(rows=vars(metric),cols=vars(site), scales="free_x") +
  labs(x="Crab ID",y="Proportion of Hashes / Reads") +
  scale_fill_manual(values=c("darkseagreen4","dodgerblue4"), name="Predator DNA") +
  theme_classic() + theme(axis.text.x=element_blank())
```

<br>

Is there a relationships between the *total reads* in a sample, and the *proportion of predator DNA*?
```{r fig.height=4, fig.width=6}
blast.dat.filter %>%
  group_by(site,crab_id,predator_rm) %>%
  summarise(nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(preads=nreads/total_reads) %>%
  filter(predator_rm==0) %>%
  dplyr::select(site,crab_id,preads, total_reads) %>%
  ggplot(aes(x=total_reads,y=preads), size=3, col="darkseagreen4") + geom_point() +
  labs(x="Total Reads in Sample",y="Proportion Reads Not Predator") +
  theme_classic()
```
<br>
<br>

### Bacterial, fungal etc. DNA

The data has the following phyla:
```{r echo=FALSE}
unique(blast.dat.filter$phylum)
```
<br>

There is also one genus of protists, and a genus of amoebas (not shown above).

These are the bacterial / fungal / amoeba / unicellular algae / (human) taxa to remove: 
```{r}
to_remove_phyla <- c("Ascomycota","Oomycota", "Basidiomycota", #fungus
                     "Tubulinea","Discosea", # amoeba
                     "Bacteroidota","Proteobacteria","Pseudomonadota", # bacteria
                     "Foraminifera") # single-celled protists
to_remove_class <- c("Cryptophyceae")   # unicellular algae
to_remove_genus <- c("Nannochloris","Micromonas","Pinnularia","Chattonellaceae","Chlamydomonadaceae","Chloropicaceae","Paralia","Tetraselmis")
to_remove_taxa <- c("Primates", "Homo sapiens","Amoebozoa sp.","uncultured organism",
                    "uncultured Jiaozhou Bay zooplankton","uncultured bacterium", "Castor canadensis")

blast.dat.filter %<>% 
  mutate(target_rm=ifelse(phylum %in% to_remove_phyla | taxon %in% to_remove_taxa | genus %in% to_remove_genus, 1, 0))
``` 
```{r echo=FALSE, message=FALSE, fig.width=10, fig.height=5}
blast.dat.filter %>%
  group_by(site,crab_id,target_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(site,crab_id,target_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=crab_id,y=proportion, fill=as.factor(target_rm))) + geom_col() + facet_grid(rows=vars(metric),cols=vars(site), scales="free_x") +
  labs(x="Crab ID",y="Proportion of Hashes / Reads") +
  scale_fill_manual(values=c("darkseagreen4","dodgerblue4"), name="Off-Target DNA") +
  theme_classic() + theme(axis.text.x=element_blank())
```
<br>

Great, an ok % of off-target hashes (around 20%) translates into a very low proportion of off-target reads.
<br>



# Save

## reads removed
Save a summary of the reason for removal for each Hash.  
```{r echo=FALSE}
summary_dat <- blast.dat.filter %>%
  mutate(retained=ifelse(missing_rm==0 & predator_rm==0 & target_rm==0, 1,0)) %>%
  pivot_longer(cols=c(predator_rm,target_rm,retained), names_to="reason") %>%
  mutate(reason=ifelse(reason=="predator_rm","predator DNA",
                       ifelse(reason=="target_rm","non-target DNA",
                              ifelse(reason=="retained","retained",NA)))) %>%
  filter(value==1) %>%
  dplyr::select(site,crab_id,Sample_name,sample_id,tech,Hash,nReads,reason,rank,taxon)

write.csv(summary_dat, here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtering_summary.csv")))
```
<br>


By MiSeq Run / site: 
```{r}
blast.dat.filter %>%
  mutate(retained=ifelse(missing_rm==0 & predator_rm==0 & target_rm==0, 1,0)) %>%
  pivot_longer(cols=c(predator_rm,target_rm,retained), names_to="reason") %>%
  mutate(reason=ifelse(reason=="predator_rm","predator DNA",
                       ifelse(reason=="target_rm","non-target DNA",
                              ifelse(reason=="retained","retained",NA)))) %>%
  filter(value==1) %>%
  group_by(MiSeqRun,site,sample_id) %>%
  mutate(totalReads=sum(nReads)) %>% ungroup() %>%
  group_by(MiSeqRun,site,sample_id,reason,totalReads) %>%
  summarise(sumReads=sum(nReads)) %>% mutate(pReads=sumReads/totalReads) %>%
  ggplot(aes(x=site, y=pReads,col=reason)) + geom_boxplot() + 
  facet_grid(cols=vars(MiSeqRun), scales="free_x") +
  labs(y="proportion of reads", title="compare MiSeq runs") +
  theme_bw() + theme(axis.text.x=element_text(angle=45,hjust=1)) 
```

```{r}
blast.dat.filter %>%
  filter(missing_rm==0 & predator_rm==0 & target_rm==0) %>%
  group_by(MiSeqRun,site,sample_id) %>%
  summarise(sumReads=sum(nReads)) %>%
  ggplot(aes(x=site, y=sumReads)) + geom_boxplot() + 
  facet_grid(cols=vars(MiSeqRun), scales="free_x") +
  labs(y="number of reads", title="compare MiSeq runs") +
  theme_bw() + theme(axis.text.x=element_text(angle=45,hjust=1)) 
```
check out the low-read samples in run 2:
```{r}
blast.dat.filter %>%
  filter(missing_rm==0 & predator_rm==0 & target_rm==0) %>%
  group_by(MiSeqRun,site,sample_id) %>%
  summarise(sumReads=sum(nReads)) %>% filter(MiSeqRun==2 & sumReads < 20)
```



## taxonomy

save only the remaining prey taxa.
```{r}
blast.dat.out <- blast.dat.filter %>%
  filter(missing_rm==0 & predator_rm==0 & target_rm==0) %>%
  dplyr::select(-"missing_rm",-"predator_rm",-"target_rm")
```
<br>

write out. 
```{r}
write_csv(blast.dat.out, here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered.csv")))
```
<br>


Did we lose any samples?
```{r}
metadat[which(!(metadat$sample_label %in% blast.dat.out$sample_id)),]
```

How many crab per site are present in the filtered blast data?
```{r}
site.totals <- blast.dat.out %>% group_by(site) %>% summarise(total_crab=length(unique(crab_id)), total_hashes=length(unique(Hash)), total_reads=sum(nReads))
site.totals %>% dplyr::select(site,total_crab)
```

## unique taxa

```{r}
dat.out <- read_csv(here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered.csv")))
```

save just the species-level IDs
```{r}
spdat <- dat.out %>% filter(rank=="species")
write_csv(spdat, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered_species.csv"))))

spdat_summary <- spdat %>% group_by(taxon,kingdom,phylum,class,order,family,genus,species) %>%
  summarise(n.crab=length(unique(crab_id)),
            n.sites=length(unique(site)),
            sites=paste(site,collapse=", "))

write_csv(spdat_summary, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered_speciesSUMMARY.csv"))))
```

keep non-specific ids in data set *only if that taxon is not represented for the given crab in the species-level data* -- for all plant / algae IDs allow for order thru genus, otherwise just through genus
```{r}
nonspec_dat <- dat.out %>% 
  filter((phylum %in% c("Chlorophyta","Rhodophyta","Streptophyta") & rank=="order") | (class %in% c("Phaeophyceae") & rank=="order") | rank %in% c("genus","family")) %>%
  mutate(dup_sample=crab_id)

nonspec_dat %<>%
  group_by(crab_id) %>%
  nest()

remove_duplicate_taxonomy <- function(x){
  tmp_sample <- as.character(unique(x[,"dup_sample"]))
  
  if(tmp_sample %in% spdat$crab_id){
    green.filter <- x %>% filter(phylum %in% c("Chlorophyta","Rhodophyta","Streptophyta") | (class %in% c("Phaeophyceae","Chrysophyceae"))) %>%
      filter(!(taxon %in% (filter(spdat,crab_id==tmp_sample))$genus) & !(taxon %in% (filter(spdat,crab_id==tmp_sample))$family) & !(taxon %in% (filter(spdat,crab_id==tmp_sample))$order))
    x.filter <- x %>% filter(!(phylum %in% c("Chlorophyta","Rhodophyta","Streptophyta")) & (!(class %in% c("Phaeophyceae","Chrysophyceae")))) %>%
      filter(!(taxon %in% (filter(spdat,crab_id==tmp_sample))$genus) & !(taxon %in% (filter(spdat,crab_id==tmp_sample))$family)) %>% 
    bind_rows(green.filter)
    return(x.filter)
    
  } else{return(x)}
}

nonspec_dat %<>% mutate(data.filter=purrr::map(data,remove_duplicate_taxonomy))

unqdat <- nonspec_dat %>% dplyr::select(-data) %>% unnest(c(crab_id,data.filter)) %>% dplyr::select(-dup_sample) %>%
  bind_rows(spdat)
```

manually remove duplicates that have two + entries of higher level taxonomy.
```{r}
duplicates_to_edit <- unqdat %>% 
  filter(rank!="species") %>% 
  dplyr::select(crab_id,Hash,taxon,rank,species,genus,family,order,class) %>% distinct() %>%
  group_by(Hash, crab_id) %>% summarise(duplicated.ranks=length(unique(rank))) %>% filter(duplicated.ranks > 1)
duplicates_to_edit
# View(left_join(duplicates_to_edit, unqdat))
```
```{r}
# fix hashes
tmpdat <- filter(unqdat, Hash=="1190ea6a61d23f996ab852bf67978413355a6b53" & crab_id %in% c("SAMT_01")) %>%
  filter(rank=="genus") %>%
  bind_rows(
    filter(unqdat, Hash=="4b6c8c35716974f06be0f76c7de6843794dc1b73" & crab_id %in% c("CLAY_13")) %>%
  filter(rank=="family")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="5434d354d98463a68a0e753c78525f1f892ec06d" & crab_id %in% c("KAY_25")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="a67e6e12eb3f3b86509f15407feb68237b2c5f83" & crab_id %in% c("MARPT_16")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="b794dcd132dede4484b9f61bb7e3810d1b42b5b4" & crab_id %in% c("MARPT_16")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="d6c968ea8f9865b7706218a113d9f92851824e7b" & crab_id %in% c("SIN_03")) %>%
  filter(rank=="genus")
  )


unqdat %<>% anti_join(dplyr::select(duplicates_to_edit, c(Hash,crab_id))) %>%
  bind_rows(tmpdat)
```

Save 
```{r}
write_csv(unqdat, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered_uniqueTaxa.csv"))))


unq_summary <- unqdat %>%
  group_by(taxon,rank,kingdom,phylum,class,order,family,genus,species) %>%
  summarise(n.crab=length(unique(crab_id)),
            n.sites=length(unique(site)),
            sites=paste0(unique(site),collapse=","),
            miseqruns=paste0(unique(MiSeqRun),collapse=","))
write_csv(unq_summary, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered_uniqueTaxaSUMMARY.csv"))))

```



# Manual filter

## update dataset

After manually checking taxon identifications against databases, I want to filter out the potential mis-identifications and missed non-target IDs. 

```{r}
manual_filter <- readxl::read_excel(path=here('data','results','lerayXT_r1-2_sample_taxonomy_filtered_uniqueTaxaSUMMARY.xlsx'),sheet='lerayXT_r1-2_final')
head(manual_filter)
```

If I need to filter out the taxon, new_rank should be "NA". the reason for removing the taxon will be in the new_taxon column. 

```{r}
filter(manual_filter, new_rank=="NA") %>% pull(new_taxon)
```

Join the dataframe to the unique taxa dataset, and filter out taxa.  Also get rid of some un-needed columns.
```{r}
unqdat_filter <- unqdat %>%
  dplyr::select(-c(19:36)) %>%
  left_join(manual_filter, by=c("taxon","rank")) %>%
  filter(new_rank!="NA")
```

Rename taxa if needed (usually to back out species to genus, if the species name is just a sample identifier from blast, or if the species is not known to be found in the Puget Sound but the genus is.)
```{r}
unqdat_filter <- unqdat_filter %>% filter(rank!=new_rank) %>%
  mutate(taxon=new_taxon,
         rank=new_rank) %>%
  bind_rows(filter(unqdat_filter,rank==new_rank))

## check some of the renamed taxa
head(filter(unqdat_filter,rank=="genus" & !is.na(species)) %>% dplyr::select(crab_id,sample_id,MiSeqRun,rank,taxon,species))
```


## remove duplicate IDs

Repeat the process above, where duplicate taxonomic identifications per crab are removed (because of a species/genus combination or similar)

First, grab species-level IDs
```{r}
# species-level IDs
spdat <- unqdat_filter %>% filter(rank=="species")

spdat_summary <- spdat %>% group_by(taxon,kingdom,phylum,class,order,family,genus,species) %>%
  summarise(n.crab=length(unique(crab_id)),
            n.sites=length(unique(site)),
            sites=paste(site,collapse=", "))

write_csv(spdat_summary, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_speciesSUMMARY.csv"))))
```

keep non-specific ids in data set *only if that taxon is not represented for the given crab in the species-level data* -- for all plant / algae IDs allow for order thru genus, otherwise just through genus
```{r}
nonspec_dat <- unqdat_filter %>% 
  filter(rank != "species") %>%
  mutate(dup_sample=crab_id)

nonspec_dat %<>%
  group_by(crab_id) %>%
  nest()

remove_duplicate_taxonomy <- function(x){
  tmp_sample <- as.character(unique(x[,"dup_sample"]))
  
  if(tmp_sample %in% spdat$crab_id){
    x.filter <- x %>%
      filter(!(taxon %in% (filter(spdat,crab_id==tmp_sample))$genus) & !(taxon %in% (filter(spdat,crab_id==tmp_sample))$family) & !(taxon %in% (filter(spdat,crab_id==tmp_sample))$order))
    return(x.filter)
    
  } else{return(x)}
}

nonspec_dat %<>% mutate(data.filter=purrr::map(data,remove_duplicate_taxonomy))

unqdat_filter2 <- nonspec_dat %>% dplyr::select(-data) %>% unnest(c(crab_id,data.filter)) %>% dplyr::select(-dup_sample) %>%
  bind_rows(spdat)
```


*********************************************
manually remove duplicates that have two + entries of higher level taxonomy.
```{r}
duplicates_to_edit <- unqdat_filter2 %>% 
  filter(rank!="species") %>% 
  dplyr::select(crab_id,Hash,taxon,rank,species,genus,family,order,class) %>% distinct() %>%
  group_by(Hash, crab_id) %>% summarise(duplicated.ranks=length(unique(rank))) %>% filter(duplicated.ranks > 1)
duplicates_to_edit
# View(left_join(duplicates_to_edit, unqdat))
```
```{r}
# fix hashes
tmpdat <- filter(unqdat, Hash=="1190ea6a61d23f996ab852bf67978413355a6b53" & crab_id %in% c("SAMT_01")) %>%
  filter(rank=="genus") %>%
  bind_rows(
    filter(unqdat, Hash=="4b6c8c35716974f06be0f76c7de6843794dc1b73" & crab_id %in% c("CLAY_13")) %>%
  filter(rank=="family")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="5434d354d98463a68a0e753c78525f1f892ec06d" & crab_id %in% c("KAY_25")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="a67e6e12eb3f3b86509f15407feb68237b2c5f83" & crab_id %in% c("MARPT_16")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="b794dcd132dede4484b9f61bb7e3810d1b42b5b4" & crab_id %in% c("MARPT_16")) %>%
  filter(rank=="genus")
  ) %>%
  bind_rows(
    filter(unqdat, Hash=="d6c968ea8f9865b7706218a113d9f92851824e7b" & crab_id %in% c("SIN_03")) %>%
  filter(rank=="genus")
  )


unqdat %<>% anti_join(dplyr::select(duplicates_to_edit, c(Hash,crab_id))) %>%
  bind_rows(tmpdat)
```


Save the final filtered taxonomy
```{r}
write_csv(unqdat_filter, (here(outdir, paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_uniqueTaxa.csv"))))
```

Save a summary of the final filtered taxonomy
```{r}
unq_filter_summary <- unqdat_filter %>%
  group_by(taxon,rank,kingdom,phylum,class,order,family,genus,species) %>%
  summarise(n.crab=length(unique(crab_id)),
            n.sites=length(unique(site)),
            sites=paste0(unique(site),collapse=","),
            miseqruns=paste0(unique(MiSeqRun),collapse=","))
write_csv(unq_filter_summary, (here('data','results', paste0(marker,"_","r",paste(run.nums,collapse="-"),"_sample_taxonomy_filtered2_uniqueTaxaSUMMARY.csv"))))
```
