---
title: "Processing Blast Output: Taxonomy Clean Up"
author: "M Fisher"
date: '2022-10-12'
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description

Clean up the taxonomic information for each crab (+technical replicate). This involves:

1. Removing predator DNA

2. Removing non-target bacteria, fungal DNA


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(ggplot2)
library(ggrepel)
library(magrittr)
```

Specify user directories
```{r}
dada_dir    <- 'data/dada2'
blast_dir   <- 'data/blast'
outdir      <- 'data/blast'
```
<br>

What are the names of the processed blast file with taxonomy from script 6?
```{r}
# hashes with unique staxids
blast_file <- 'lerayXT_r1_sample_blast_lca_taxonomy.csv'
```
<br>

What is the location / name of the metadata file for this sequencing run?
```{r}
run_metadat_file <- 'data/DCRB_Run1_samples.csv'
```
<br>

And the run number and the marker
```{r}
run.num <- 1
marker <- 'lerayXT'
```
<br>

# Read in data

```{r read in asvs1, include=FALSE}
blast.dat <- read_csv(here(blast_dir, blast_file))
metadat <- read_csv(here(run_metadat_file))
```
<br>

Join the blast data to metadata information. Remove positive and negative controls (their hashes were removed in the filtering / denoising step before blasting.)
```{r}
blast.dat.filter <- blast.dat %>%
  left_join(metadat %>% unite(col="sample_id", c(Sample_name,Sample_label), sep="-", remove=FALSE), by=c("sample_id","Sample_name","Locus")) %>%
  filter(!(grepl("Control",sample_id))) %>%
  separate(Sample_label, into=c("temp1","temp2","temp3"), sep="-", remove=FALSE) %>% unite(col="crab_id",c(temp1,temp2)) %>% dplyr::select(-temp3)
```
<br>

Do all crab have three replicates present? *Yay! all of them.*
```{r}
blast.dat.filter %>% group_by(crab_id) %>% summarise(nrep=length(unique(Replicate))) %>% filter(nrep < 3)
```
<br>

How many crab per site are present in the blast data?
```{r}
site.totals <- blast.dat.filter %>% group_by(Site) %>% summarise(total_crab=length(unique(crab_id)), total_hashes=length(unique(Hash)), total_reads=sum(nReads))
site.totals %>% dplyr::select(Site,total_crab)
```
<br>

# Identify Taxonomy for Removal

To easier track how much data is being removed, I'm going to classify each hash as (1) to remove, and (2) reason for removal. then in the next section, I'll filter the data and check out the prey taxa. 

### Missing or Low Resolution Taxonomy

Remove any hashes that don't have taxonomy assigned, *OR* don't have a taxonomic rank at the order level or lower.
```{r}
blast.dat.filter %<>% mutate(missing_rm=ifelse(is.na(rank) | rank %in% c("kingdom","phylum","class"), 1,0))
```
```{r echo=FALSE, message=FALSE}
blast.dat.filter %>% group_by(Site,missing_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(site.totals,by="Site") %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(Site,missing_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=Site,y=proportion, fill=as.factor(missing_rm))) + geom_col() + facet_wrap(~metric) +
  labs(x="Site",y="Proportion of Hashes / Reads") +
  scale_fill_manual(values=c("darkseagreen4","black"), name="Missing Tax") +
  theme_classic()
```
<br>

Let's get rid of those hashes missing taxonomy.
```{r}
blast.dat.filter %<>% filter(missing_rm==0)
```
<br>

calculate per crab metrics
```{r}
crab.totals <- blast.dat.filter %>%
  group_by(crab_id) %>% summarise(total_hashes=length(unique(Hash)), total_reads=sum(nReads))
```
<br>

### Predator DNA

Any Dungeness crab DNA will be marked as 'predator' DNA and removed. 

- Metacarcinus magister (species)
- Metacarcinus (genus)
<br>

```{r}
blast.dat.filter %<>% 
  mutate(predator_rm=ifelse(missing_rm==0 & (taxon =="Metacarcinus magister" | taxon =="Metacarcinus"), 1, 0))
```
```{r echo=FALSE, message=FALSE, fig.width=10, fig.height=5}
blast.dat.filter %>% filter(missing_rm==0) %>%
  group_by(Site,crab_id,predator_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(Site,crab_id,predator_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=crab_id,y=proportion, fill=as.factor(predator_rm))) + geom_col() + facet_grid(cols=vars(metric),rows=vars(Site), drop=TRUE) +
  labs(x="Crab ID",y="Proportion of Hashes / Reads") +
  scale_fill_manual(values=c("darkseagreen4","dodgerblue4"), name="Predator DNA") +
  theme_classic() + theme(axis.text.x=element_blank())
```

<br>

Is there a relationships between the *total reads* in a sample, and the *proportion of predator DNA*?
```{r fig.height=4, fig.width=6}
blast.dat.filter %>% filter(missing_rm==0) %>%
  group_by(Site,crab_id,predator_rm) %>%
  summarise(nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(preads=nreads/total_reads) %>%
  filter(predator_rm==0) %>%
  dplyr::select(Site,crab_id,preads, total_reads) %>%
  ggplot(aes(x=total_reads,y=preads), size=3, col="darkseagreen4") + geom_point() +
  labs(x="Total Reads in Sample",y="Proportion Reads Not Predator") +
  theme_classic()
```
<br>
<br>

### Bacterial, fungal etc. DNA

The data has the following phyla:
```{r echo=FALSE}
unique(blast.dat.filter$phylum)
```
<br>

There is also one genus of protists, and a genus of amoebas (not shown above).

These are the bacterial / fungal / amoeba / (human) taxa to remove: 
```{r}
to_remove_phyla <- c("Ascomycota", "Bacteroidota","Basidiomycota","Discosea","Oomycota","Proteobacteria","Tubulinea")
to_remove_taxa <- c("Primates", "Homo sapiens","Amoebozoa sp.")

blast.dat.filter %<>% 
  mutate(target_rm=ifelse(predator_rm==0 & (phylum %in% to_remove_phyla | taxon %in% to_remove_taxa), 1, 0))
``` 
```{r echo=FALSE, message=FALSE, fig.width=10, fig.height=5}
blast.dat.filter %>%
  group_by(Site,crab_id,target_rm) %>%
  summarise(nhashes=length(unique(Hash)),nreads=sum(nReads)) %>%
  left_join(crab.totals,by=c("crab_id")) %>%
  mutate(phashes=nhashes/total_hashes,
         preads=nreads/total_reads) %>%
  dplyr::select(Site,crab_id,target_rm,phashes,preads) %>%
  pivot_longer(cols=c(phashes,preads),names_to="metric",values_to="proportion") %>%
  ggplot(aes(x=crab_id,y=proportion, fill=as.factor(target_rm))) + geom_col() + facet_grid(cols=vars(metric),rows=vars(Site), drop=TRUE, scales="free") +
  labs(x="Crab ID",y="Proportion of Hashes / Reads") +
  scale_fill_manual(values=c("darkseagreen4","dodgerblue4"), name="Off-Target DNA") +
  theme_classic() + theme(axis.text.x=element_blank())
```
<br>

Great, an ok % of off-target hashes (around 20%) translates into a very low proportion of off-target reads.
<br>


# Save

## reads removed
Save a summary of the reason for removal for each Hash.  

Look at the proportion of reads in each 'removal' category, for each sample.
```{r echo=FALSE}
summary_dat <- blast.dat.filter %>%
  mutate(retained=ifelse(missing_rm==0 & predator_rm==0 & target_rm==0, 1,0)) %>%
  pivot_longer(cols=c(predator_rm,target_rm,retained), names_to="reason") %>%
  mutate(reason=ifelse(reason=="predator_rm","predator DNA",
                       ifelse(reason=="target_rm","non-target DNA",
                              ifelse(reason=="retained","retained",NA)))) %>%
  filter(value==1) %>%
  dplyr::select(Site,crab_id,Sample_name,sample_id,Replicate,Hash,nReads,reason,rank,taxon)

write.csv(summary_dat, here(outdir, paste0(marker,"_","r",run.num,"_sample_taxonomy_filtering_summary.csv")))
```
<br>

## taxonomy

save only the remaining prey taxa.
```{r}
blast.dat.out <- blast.dat.filter %>%
  filter(missing_rm==0 & predator_rm==0 & target_rm==0) %>%
  dplyr::select(-"missing_rm",-"predator_rm",-"target_rm")
```
<br>

write out. 
```{r}
write_csv(blast.dat.out, here(outdir, paste0(marker,"_","r",run.num,"_sample_taxonomy_filtered.csv")))
```
<br>








